{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "name": "Drug_Binding_Miniproject.ipynb",
      "provenance": [],
      "collapsed_sections": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.6.5"
    },
    "toc": {
      "base_numbering": 1,
      "nav_menu": {},
      "number_sections": true,
      "sideBar": true,
      "skip_h1_title": false,
      "title_cell": "Table of Contents",
      "title_sidebar": "Contents",
      "toc_cell": false,
      "toc_position": {},
      "toc_section_display": true,
      "toc_window_display": false
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "XN6dxHwuSFXW"
      },
      "source": [
        "# AM216 Mini-Project: Drug target interaction and COVID-19\n",
        "\n",
        "Rapid determination of whether a candidate small molecule will bind to a particular target receptor protein remains a stumbling block in drug discovery.  If you can find a small molecule that binds to a relevant protein and modifies its function, this molecule can serve as a drug.   This is an **inverse problem**: Given a protein, please find a small molecule that can bind to it.  \n",
        "\n",
        "One way of solving this inverse problem would be to use molecular dynamics: Create a molecular model of a protein. Create a molecular model of all possible small molecule drugs that can bind to the protein. Simulate the interactions between the small molecules and the proteins. Find the molecules that work.\n",
        "\n",
        "Or-you might be more mathematical and try to solve this as a straight-up optimization problem: Find the small molecule that optimally binds to the binding pocket of a given drug.\n",
        "\n",
        "Although conceptually appealing, this methodology is impossible: The major bottleneck is that we do not know and cannot accurately represent the interactions between small molecules and proteins.  The potentials are empirical with fitting parameters--they cannot be found from first principles and despite decades of work, we simply do not have good representations for them. For that reason, physical computation has been of limited utility for drug discovery.  \n",
        "\n",
        "We should mention that a secondary bottleneck is the computing power that would be required for this search -- but this is something that could be likely sorted out if it weren't for the first bottleneck.\n",
        "\n",
        "An alternative approach is to use a *data driven approach*. Instead of representing the physics, lets make a list of all known proteins and all small molecules that binds to each protein.  Lets then design a data driven way of associating small molecules to proteins, representing binding.  \n",
        "\n",
        "Note that for a data driven approach to succeed, the key step is to **find ways of representing the small molecule and the protein that make it possible to find the patterns in the existing datasets.**   Representation is the key problem.  Presumably, the best way to represent the proteins and the small molecules is to use the physics and biochemistry of what happens when binding occurs -- as by doing this, we will focus on features that are relevant.  Finding the right feature representations for this type of problem is a very active area of research.\n",
        "\n",
        "In this miniproject, you will dip your toe into the pond by studying a canonical problem, the binding affinity of small molecules against target proteins.  \n",
        "\n",
        "Given all that is happening in the world right now, it seems appropriate to assign you the additional mission is to use the trained model to identify drugs that could bind to the main protease$^*$ protein of COVID-19 and prevent the production of viral enzyme. \n",
        "\n",
        "To help you in this mission, you are given KIBA (\"Kinase Inhibitor Bioassay\") dataset which contains 2111 drugs and 229 proteins with total of 118254 drug-protein binding affinity scores. Those are stored in `ligands_can.txt`, `proteins.txt`, and `Y`, respectively. The rows of `Y` index drugs and the columns index proteins. The element of `Y` is a real-valued binding score that takes into account dissociation constant $K_d$, inhibition constant $K_i$, and half maximal inhibitory concentration $IC_{50}$. Refer to [original paper](https://pubs.acs.org/doi/pdf/10.1021/ci400709d) for more detail. We also provide splitted indices for 5-fold cross validation in `train_fold_setting.txt` and testset indices in `test_fold_setting.txt`. Lastly, `6Y84_A.fasta.txt` contains the amino acid sequence of COVID-19 protease.\n",
        "\n",
        "You will evalulate the performance of your model using MSE and Corcordance Index (CI), which is slightly relaxed metric that measures whether the predicted binding affinity val- ues of two random drug–target pairs were predicted in the same order as their true values were.\n",
        "\n",
        "*: Note that there are many other drug targets for COVID-19 virus. [This page](\n",
        "https://www.guidetopharmacology.org/coronavirus.jsp) gives you a good summary. \n",
        "\n",
        "To help you get started, we will first introduce you to Rdkit and DeepChem packages."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "wVhNjK4kSFXZ"
      },
      "source": [
        "## 1. Basic Cheminformatics with Rdkit and DeepChem\n",
        "\n",
        "For this project, you will most likely heavily rely on Rdkit and Deepchem. You can install DeepChem along with Rdkit following instruction [here](https://deepchem.io). If you are using Colab, run the following code block.\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "G7fQCZJ8q9Je",
        "outputId": "341296ca-176c-45ce-abc6-1e452ca9c2b4"
      },
      "source": [
        "%tensorflow_version 1.x"
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "TensorFlow 1.x selected.\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ps7x414nDcHu",
        "outputId": "eb41268f-6da9-4880-a1d1-ab2c71c40069"
      },
      "source": [
        "# This may take a few minutes to run!\n",
        "!wget -c https://repo.anaconda.com/archive/Anaconda3-2019.10-Linux-x86_64.sh\n",
        "!chmod +x Anaconda3-2019.10-Linux-x86_64.sh\n",
        "!bash ./Anaconda3-2019.10-Linux-x86_64.sh -b -f -p /usr/local\n",
        "!conda install -y -c deepchem -c rdkit -c conda-forge -c omnia deepchem-gpu=2.3.0\n",
        "import sys\n",
        "sys.path.append('/usr/local/lib/python3.7/site-packages/')\n",
        "import deepchem as dc"
      ],
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-04-15 19:48:54--  https://repo.anaconda.com/archive/Anaconda3-2019.10-Linux-x86_64.sh\n",
            "Resolving repo.anaconda.com (repo.anaconda.com)... 104.16.131.3, 104.16.130.3, 2606:4700::6810:8303, ...\n",
            "Connecting to repo.anaconda.com (repo.anaconda.com)|104.16.131.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 530308481 (506M) [application/x-sh]\n",
            "Saving to: ‘Anaconda3-2019.10-Linux-x86_64.sh’\n",
            "\n",
            "Anaconda3-2019.10-L 100%[===================>] 505.74M   163MB/s    in 3.1s    \n",
            "\n",
            "2021-04-15 19:48:58 (163 MB/s) - ‘Anaconda3-2019.10-Linux-x86_64.sh’ saved [530308481/530308481]\n",
            "\n",
            "PREFIX=/usr/local\n",
            "Unpacking payload ...\n",
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Solving environment: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - _ipyw_jlab_nb_ext_conf==0.1.0=py37_0\n",
            "    - _libgcc_mutex==0.1=main\n",
            "    - alabaster==0.7.12=py37_0\n",
            "    - anaconda-client==1.7.2=py37_0\n",
            "    - anaconda-navigator==1.9.7=py37_0\n",
            "    - anaconda-project==0.8.3=py_0\n",
            "    - anaconda==2019.10=py37_0\n",
            "    - asn1crypto==1.0.1=py37_0\n",
            "    - astroid==2.3.1=py37_0\n",
            "    - astropy==3.2.2=py37h7b6447c_0\n",
            "    - atomicwrites==1.3.0=py37_1\n",
            "    - attrs==19.2.0=py_0\n",
            "    - babel==2.7.0=py_0\n",
            "    - backcall==0.1.0=py37_0\n",
            "    - backports.functools_lru_cache==1.5=py_2\n",
            "    - backports.os==0.1.1=py37_0\n",
            "    - backports.shutil_get_terminal_size==1.0.0=py37_2\n",
            "    - backports.tempfile==1.0=py_1\n",
            "    - backports.weakref==1.0.post1=py_1\n",
            "    - backports==1.0=py_2\n",
            "    - beautifulsoup4==4.8.0=py37_0\n",
            "    - bitarray==1.0.1=py37h7b6447c_0\n",
            "    - bkcharts==0.2=py37_0\n",
            "    - blas==1.0=mkl\n",
            "    - bleach==3.1.0=py37_0\n",
            "    - blosc==1.16.3=hd408876_0\n",
            "    - bokeh==1.3.4=py37_0\n",
            "    - boto==2.49.0=py37_0\n",
            "    - bottleneck==1.2.1=py37h035aef0_1\n",
            "    - bzip2==1.0.8=h7b6447c_0\n",
            "    - ca-certificates==2019.8.28=0\n",
            "    - cairo==1.14.12=h8948797_3\n",
            "    - certifi==2019.9.11=py37_0\n",
            "    - cffi==1.12.3=py37h2e261b9_0\n",
            "    - chardet==3.0.4=py37_1003\n",
            "    - click==7.0=py37_0\n",
            "    - cloudpickle==1.2.2=py_0\n",
            "    - clyent==1.2.2=py37_1\n",
            "    - colorama==0.4.1=py37_0\n",
            "    - conda-build==3.18.9=py37_3\n",
            "    - conda-env==2.6.0=1\n",
            "    - conda-package-handling==1.6.0=py37h7b6447c_0\n",
            "    - conda-verify==3.4.2=py_1\n",
            "    - conda==4.7.12=py37_0\n",
            "    - contextlib2==0.6.0=py_0\n",
            "    - cryptography==2.7=py37h1ba5d50_0\n",
            "    - curl==7.65.3=hbc83047_0\n",
            "    - cycler==0.10.0=py37_0\n",
            "    - cython==0.29.13=py37he6710b0_0\n",
            "    - cytoolz==0.10.0=py37h7b6447c_0\n",
            "    - dask-core==2.5.2=py_0\n",
            "    - dask==2.5.2=py_0\n",
            "    - dbus==1.13.6=h746ee38_0\n",
            "    - decorator==4.4.0=py37_1\n",
            "    - defusedxml==0.6.0=py_0\n",
            "    - distributed==2.5.2=py_0\n",
            "    - docutils==0.15.2=py37_0\n",
            "    - entrypoints==0.3=py37_0\n",
            "    - et_xmlfile==1.0.1=py37_0\n",
            "    - expat==2.2.6=he6710b0_0\n",
            "    - fastcache==1.1.0=py37h7b6447c_0\n",
            "    - filelock==3.0.12=py_0\n",
            "    - flask==1.1.1=py_0\n",
            "    - fontconfig==2.13.0=h9420a91_0\n",
            "    - freetype==2.9.1=h8a8886c_1\n",
            "    - fribidi==1.0.5=h7b6447c_0\n",
            "    - fsspec==0.5.2=py_0\n",
            "    - future==0.17.1=py37_0\n",
            "    - get_terminal_size==1.0.0=haa9412d_0\n",
            "    - gevent==1.4.0=py37h7b6447c_0\n",
            "    - glib==2.56.2=hd408876_0\n",
            "    - glob2==0.7=py_0\n",
            "    - gmp==6.1.2=h6c8ec71_1\n",
            "    - gmpy2==2.0.8=py37h10f8cd9_2\n",
            "    - graphite2==1.3.13=h23475e2_0\n",
            "    - greenlet==0.4.15=py37h7b6447c_0\n",
            "    - gst-plugins-base==1.14.0=hbbd80ab_1\n",
            "    - gstreamer==1.14.0=hb453b48_1\n",
            "    - h5py==2.9.0=py37h7918eee_0\n",
            "    - harfbuzz==1.8.8=hffaf4a1_0\n",
            "    - hdf5==1.10.4=hb1b8bf9_0\n",
            "    - heapdict==1.0.1=py_0\n",
            "    - html5lib==1.0.1=py37_0\n",
            "    - icu==58.2=h9c2bf20_1\n",
            "    - idna==2.8=py37_0\n",
            "    - imageio==2.6.0=py37_0\n",
            "    - imagesize==1.1.0=py37_0\n",
            "    - importlib_metadata==0.23=py37_0\n",
            "    - intel-openmp==2019.4=243\n",
            "    - ipykernel==5.1.2=py37h39e3cac_0\n",
            "    - ipython==7.8.0=py37h39e3cac_0\n",
            "    - ipython_genutils==0.2.0=py37_0\n",
            "    - ipywidgets==7.5.1=py_0\n",
            "    - isort==4.3.21=py37_0\n",
            "    - itsdangerous==1.1.0=py37_0\n",
            "    - jbig==2.1=hdba287a_0\n",
            "    - jdcal==1.4.1=py_0\n",
            "    - jedi==0.15.1=py37_0\n",
            "    - jeepney==0.4.1=py_0\n",
            "    - jinja2==2.10.3=py_0\n",
            "    - joblib==0.13.2=py37_0\n",
            "    - jpeg==9b=h024ee3a_2\n",
            "    - json5==0.8.5=py_0\n",
            "    - jsonschema==3.0.2=py37_0\n",
            "    - jupyter==1.0.0=py37_7\n",
            "    - jupyter_client==5.3.3=py37_1\n",
            "    - jupyter_console==6.0.0=py37_0\n",
            "    - jupyter_core==4.5.0=py_0\n",
            "    - jupyterlab==1.1.4=pyhf63ae98_0\n",
            "    - jupyterlab_server==1.0.6=py_0\n",
            "    - keyring==18.0.0=py37_0\n",
            "    - kiwisolver==1.1.0=py37he6710b0_0\n",
            "    - krb5==1.16.1=h173b8e3_7\n",
            "    - lazy-object-proxy==1.4.2=py37h7b6447c_0\n",
            "    - libarchive==3.3.3=h5d8350f_5\n",
            "    - libcurl==7.65.3=h20c2e04_0\n",
            "    - libedit==3.1.20181209=hc058e9b_0\n",
            "    - libffi==3.2.1=hd88cf55_4\n",
            "    - libgcc-ng==9.1.0=hdf63c60_0\n",
            "    - libgfortran-ng==7.3.0=hdf63c60_0\n",
            "    - liblief==0.9.0=h7725739_2\n",
            "    - libpng==1.6.37=hbc83047_0\n",
            "    - libsodium==1.0.16=h1bed415_0\n",
            "    - libssh2==1.8.2=h1ba5d50_0\n",
            "    - libstdcxx-ng==9.1.0=hdf63c60_0\n",
            "    - libtiff==4.0.10=h2733197_2\n",
            "    - libtool==2.4.6=h7b6447c_5\n",
            "    - libuuid==1.0.3=h1bed415_2\n",
            "    - libxcb==1.13=h1bed415_1\n",
            "    - libxml2==2.9.9=hea5a465_1\n",
            "    - libxslt==1.1.33=h7d1a2b0_0\n",
            "    - llvmlite==0.29.0=py37hd408876_0\n",
            "    - locket==0.2.0=py37_1\n",
            "    - lxml==4.4.1=py37hefd8a0e_0\n",
            "    - lz4-c==1.8.1.2=h14c3975_0\n",
            "    - lzo==2.10=h49e0be7_2\n",
            "    - markupsafe==1.1.1=py37h7b6447c_0\n",
            "    - matplotlib==3.1.1=py37h5429711_0\n",
            "    - mccabe==0.6.1=py37_1\n",
            "    - mistune==0.8.4=py37h7b6447c_0\n",
            "    - mkl-service==2.3.0=py37he904b0f_0\n",
            "    - mkl==2019.4=243\n",
            "    - mkl_fft==1.0.14=py37ha843d7b_0\n",
            "    - mkl_random==1.1.0=py37hd6b4f25_0\n",
            "    - mock==3.0.5=py37_0\n",
            "    - more-itertools==7.2.0=py37_0\n",
            "    - mpc==1.1.0=h10f8cd9_1\n",
            "    - mpfr==4.0.1=hdf1c602_3\n",
            "    - mpmath==1.1.0=py37_0\n",
            "    - msgpack-python==0.6.1=py37hfd86e86_1\n",
            "    - multipledispatch==0.6.0=py37_0\n",
            "    - navigator-updater==0.2.1=py37_0\n",
            "    - nbconvert==5.6.0=py37_1\n",
            "    - nbformat==4.4.0=py37_0\n",
            "    - ncurses==6.1=he6710b0_1\n",
            "    - networkx==2.3=py_0\n",
            "    - nltk==3.4.5=py37_0\n",
            "    - nose==1.3.7=py37_2\n",
            "    - notebook==6.0.1=py37_0\n",
            "    - numba==0.45.1=py37h962f231_0\n",
            "    - numexpr==2.7.0=py37h9e4a6bb_0\n",
            "    - numpy-base==1.17.2=py37hde5b4d6_0\n",
            "    - numpy==1.17.2=py37haad9e8e_0\n",
            "    - numpydoc==0.9.1=py_0\n",
            "    - olefile==0.46=py37_0\n",
            "    - openpyxl==3.0.0=py_0\n",
            "    - openssl==1.1.1d=h7b6447c_2\n",
            "    - packaging==19.2=py_0\n",
            "    - pandas==0.25.1=py37he6710b0_0\n",
            "    - pandoc==2.2.3.2=0\n",
            "    - pandocfilters==1.4.2=py37_1\n",
            "    - pango==1.42.4=h049681c_0\n",
            "    - parso==0.5.1=py_0\n",
            "    - partd==1.0.0=py_0\n",
            "    - patchelf==0.9=he6710b0_3\n",
            "    - path.py==12.0.1=py_0\n",
            "    - pathlib2==2.3.5=py37_0\n",
            "    - patsy==0.5.1=py37_0\n",
            "    - pcre==8.43=he6710b0_0\n",
            "    - pep8==1.7.1=py37_0\n",
            "    - pexpect==4.7.0=py37_0\n",
            "    - pickleshare==0.7.5=py37_0\n",
            "    - pillow==6.2.0=py37h34e0f95_0\n",
            "    - pip==19.2.3=py37_0\n",
            "    - pixman==0.38.0=h7b6447c_0\n",
            "    - pkginfo==1.5.0.1=py37_0\n",
            "    - pluggy==0.13.0=py37_0\n",
            "    - ply==3.11=py37_0\n",
            "    - prometheus_client==0.7.1=py_0\n",
            "    - prompt_toolkit==2.0.10=py_0\n",
            "    - psutil==5.6.3=py37h7b6447c_0\n",
            "    - ptyprocess==0.6.0=py37_0\n",
            "    - py-lief==0.9.0=py37h7725739_2\n",
            "    - py==1.8.0=py37_0\n",
            "    - pycodestyle==2.5.0=py37_0\n",
            "    - pycosat==0.6.3=py37h14c3975_0\n",
            "    - pycparser==2.19=py37_0\n",
            "    - pycrypto==2.6.1=py37h14c3975_9\n",
            "    - pycurl==7.43.0.3=py37h1ba5d50_0\n",
            "    - pyflakes==2.1.1=py37_0\n",
            "    - pygments==2.4.2=py_0\n",
            "    - pylint==2.4.2=py37_0\n",
            "    - pyodbc==4.0.27=py37he6710b0_0\n",
            "    - pyopenssl==19.0.0=py37_0\n",
            "    - pyparsing==2.4.2=py_0\n",
            "    - pyqt==5.9.2=py37h05f1152_2\n",
            "    - pyrsistent==0.15.4=py37h7b6447c_0\n",
            "    - pysocks==1.7.1=py37_0\n",
            "    - pytables==3.5.2=py37h71ec239_1\n",
            "    - pytest-arraydiff==0.3=py37h39e3cac_0\n",
            "    - pytest-astropy==0.5.0=py37_0\n",
            "    - pytest-doctestplus==0.4.0=py_0\n",
            "    - pytest-openfiles==0.4.0=py_0\n",
            "    - pytest-remotedata==0.3.2=py37_0\n",
            "    - pytest==5.2.1=py37_0\n",
            "    - python-dateutil==2.8.0=py37_0\n",
            "    - python-libarchive-c==2.8=py37_13\n",
            "    - python==3.7.4=h265db76_1\n",
            "    - pytz==2019.3=py_0\n",
            "    - pywavelets==1.0.3=py37hdd07704_1\n",
            "    - pyyaml==5.1.2=py37h7b6447c_0\n",
            "    - pyzmq==18.1.0=py37he6710b0_0\n",
            "    - qt==5.9.7=h5867ecd_1\n",
            "    - qtawesome==0.6.0=py_0\n",
            "    - qtconsole==4.5.5=py_0\n",
            "    - qtpy==1.9.0=py_0\n",
            "    - readline==7.0=h7b6447c_5\n",
            "    - requests==2.22.0=py37_0\n",
            "    - ripgrep==0.10.0=hc07d326_0\n",
            "    - rope==0.14.0=py_0\n",
            "    - ruamel_yaml==0.15.46=py37h14c3975_0\n",
            "    - scikit-image==0.15.0=py37he6710b0_0\n",
            "    - scikit-learn==0.21.3=py37hd81dba3_0\n",
            "    - scipy==1.3.1=py37h7c811a0_0\n",
            "    - seaborn==0.9.0=py37_0\n",
            "    - secretstorage==3.1.1=py37_0\n",
            "    - send2trash==1.5.0=py37_0\n",
            "    - setuptools==41.4.0=py37_0\n",
            "    - simplegeneric==0.8.1=py37_2\n",
            "    - singledispatch==3.4.0.3=py37_0\n",
            "    - sip==4.19.8=py37hf484d3e_0\n",
            "    - six==1.12.0=py37_0\n",
            "    - snappy==1.1.7=hbae5bb6_3\n",
            "    - snowballstemmer==2.0.0=py_0\n",
            "    - sortedcollections==1.1.2=py37_0\n",
            "    - sortedcontainers==2.1.0=py37_0\n",
            "    - soupsieve==1.9.3=py37_0\n",
            "    - sphinx==2.2.0=py_0\n",
            "    - sphinxcontrib-applehelp==1.0.1=py_0\n",
            "    - sphinxcontrib-devhelp==1.0.1=py_0\n",
            "    - sphinxcontrib-htmlhelp==1.0.2=py_0\n",
            "    - sphinxcontrib-jsmath==1.0.1=py_0\n",
            "    - sphinxcontrib-qthelp==1.0.2=py_0\n",
            "    - sphinxcontrib-serializinghtml==1.1.3=py_0\n",
            "    - sphinxcontrib-websupport==1.1.2=py_0\n",
            "    - sphinxcontrib==1.0=py37_1\n",
            "    - spyder-kernels==0.5.2=py37_0\n",
            "    - spyder==3.3.6=py37_0\n",
            "    - sqlalchemy==1.3.9=py37h7b6447c_0\n",
            "    - sqlite==3.30.0=h7b6447c_0\n",
            "    - statsmodels==0.10.1=py37hdd07704_0\n",
            "    - sympy==1.4=py37_0\n",
            "    - tbb==2019.4=hfd86e86_0\n",
            "    - tblib==1.4.0=py_0\n",
            "    - terminado==0.8.2=py37_0\n",
            "    - testpath==0.4.2=py37_0\n",
            "    - tk==8.6.8=hbc83047_0\n",
            "    - toolz==0.10.0=py_0\n",
            "    - tornado==6.0.3=py37h7b6447c_0\n",
            "    - tqdm==4.36.1=py_0\n",
            "    - traitlets==4.3.3=py37_0\n",
            "    - unicodecsv==0.14.1=py37_0\n",
            "    - unixodbc==2.3.7=h14c3975_0\n",
            "    - urllib3==1.24.2=py37_0\n",
            "    - wcwidth==0.1.7=py37_0\n",
            "    - webencodings==0.5.1=py37_1\n",
            "    - werkzeug==0.16.0=py_0\n",
            "    - wheel==0.33.6=py37_0\n",
            "    - widgetsnbextension==3.5.1=py37_0\n",
            "    - wrapt==1.11.2=py37h7b6447c_0\n",
            "    - wurlitzer==1.0.3=py37_0\n",
            "    - xlrd==1.2.0=py37_0\n",
            "    - xlsxwriter==1.2.1=py_0\n",
            "    - xlwt==1.3.0=py37_0\n",
            "    - xz==5.2.4=h14c3975_4\n",
            "    - yaml==0.1.7=had09818_2\n",
            "    - zeromq==4.3.1=he6710b0_3\n",
            "    - zict==1.0.0=py_0\n",
            "    - zipp==0.6.0=py_0\n",
            "    - zlib==1.2.11=h7b6447c_3\n",
            "    - zstd==1.3.7=h0b5b093_0\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _ipyw_jlab_nb_ext~ pkgs/main/linux-64::_ipyw_jlab_nb_ext_conf-0.1.0-py37_0\n",
            "  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main\n",
            "  alabaster          pkgs/main/linux-64::alabaster-0.7.12-py37_0\n",
            "  anaconda           pkgs/main/linux-64::anaconda-2019.10-py37_0\n",
            "  anaconda-client    pkgs/main/linux-64::anaconda-client-1.7.2-py37_0\n",
            "  anaconda-navigator pkgs/main/linux-64::anaconda-navigator-1.9.7-py37_0\n",
            "  anaconda-project   pkgs/main/noarch::anaconda-project-0.8.3-py_0\n",
            "  asn1crypto         pkgs/main/linux-64::asn1crypto-1.0.1-py37_0\n",
            "  astroid            pkgs/main/linux-64::astroid-2.3.1-py37_0\n",
            "  astropy            pkgs/main/linux-64::astropy-3.2.2-py37h7b6447c_0\n",
            "  atomicwrites       pkgs/main/linux-64::atomicwrites-1.3.0-py37_1\n",
            "  attrs              pkgs/main/noarch::attrs-19.2.0-py_0\n",
            "  babel              pkgs/main/noarch::babel-2.7.0-py_0\n",
            "  backcall           pkgs/main/linux-64::backcall-0.1.0-py37_0\n",
            "  backports          pkgs/main/noarch::backports-1.0-py_2\n",
            "  backports.functoo~ pkgs/main/noarch::backports.functools_lru_cache-1.5-py_2\n",
            "  backports.os       pkgs/main/linux-64::backports.os-0.1.1-py37_0\n",
            "  backports.shutil_~ pkgs/main/linux-64::backports.shutil_get_terminal_size-1.0.0-py37_2\n",
            "  backports.tempfile pkgs/main/noarch::backports.tempfile-1.0-py_1\n",
            "  backports.weakref  pkgs/main/noarch::backports.weakref-1.0.post1-py_1\n",
            "  beautifulsoup4     pkgs/main/linux-64::beautifulsoup4-4.8.0-py37_0\n",
            "  bitarray           pkgs/main/linux-64::bitarray-1.0.1-py37h7b6447c_0\n",
            "  bkcharts           pkgs/main/linux-64::bkcharts-0.2-py37_0\n",
            "  blas               pkgs/main/linux-64::blas-1.0-mkl\n",
            "  bleach             pkgs/main/linux-64::bleach-3.1.0-py37_0\n",
            "  blosc              pkgs/main/linux-64::blosc-1.16.3-hd408876_0\n",
            "  bokeh              pkgs/main/linux-64::bokeh-1.3.4-py37_0\n",
            "  boto               pkgs/main/linux-64::boto-2.49.0-py37_0\n",
            "  bottleneck         pkgs/main/linux-64::bottleneck-1.2.1-py37h035aef0_1\n",
            "  bzip2              pkgs/main/linux-64::bzip2-1.0.8-h7b6447c_0\n",
            "  ca-certificates    pkgs/main/linux-64::ca-certificates-2019.8.28-0\n",
            "  cairo              pkgs/main/linux-64::cairo-1.14.12-h8948797_3\n",
            "  certifi            pkgs/main/linux-64::certifi-2019.9.11-py37_0\n",
            "  cffi               pkgs/main/linux-64::cffi-1.12.3-py37h2e261b9_0\n",
            "  chardet            pkgs/main/linux-64::chardet-3.0.4-py37_1003\n",
            "  click              pkgs/main/linux-64::click-7.0-py37_0\n",
            "  cloudpickle        pkgs/main/noarch::cloudpickle-1.2.2-py_0\n",
            "  clyent             pkgs/main/linux-64::clyent-1.2.2-py37_1\n",
            "  colorama           pkgs/main/linux-64::colorama-0.4.1-py37_0\n",
            "  conda              pkgs/main/linux-64::conda-4.7.12-py37_0\n",
            "  conda-build        pkgs/main/linux-64::conda-build-3.18.9-py37_3\n",
            "  conda-env          pkgs/main/linux-64::conda-env-2.6.0-1\n",
            "  conda-package-han~ pkgs/main/linux-64::conda-package-handling-1.6.0-py37h7b6447c_0\n",
            "  conda-verify       pkgs/main/noarch::conda-verify-3.4.2-py_1\n",
            "  contextlib2        pkgs/main/noarch::contextlib2-0.6.0-py_0\n",
            "  cryptography       pkgs/main/linux-64::cryptography-2.7-py37h1ba5d50_0\n",
            "  curl               pkgs/main/linux-64::curl-7.65.3-hbc83047_0\n",
            "  cycler             pkgs/main/linux-64::cycler-0.10.0-py37_0\n",
            "  cython             pkgs/main/linux-64::cython-0.29.13-py37he6710b0_0\n",
            "  cytoolz            pkgs/main/linux-64::cytoolz-0.10.0-py37h7b6447c_0\n",
            "  dask               pkgs/main/noarch::dask-2.5.2-py_0\n",
            "  dask-core          pkgs/main/noarch::dask-core-2.5.2-py_0\n",
            "  dbus               pkgs/main/linux-64::dbus-1.13.6-h746ee38_0\n",
            "  decorator          pkgs/main/linux-64::decorator-4.4.0-py37_1\n",
            "  defusedxml         pkgs/main/noarch::defusedxml-0.6.0-py_0\n",
            "  distributed        pkgs/main/noarch::distributed-2.5.2-py_0\n",
            "  docutils           pkgs/main/linux-64::docutils-0.15.2-py37_0\n",
            "  entrypoints        pkgs/main/linux-64::entrypoints-0.3-py37_0\n",
            "  et_xmlfile         pkgs/main/linux-64::et_xmlfile-1.0.1-py37_0\n",
            "  expat              pkgs/main/linux-64::expat-2.2.6-he6710b0_0\n",
            "  fastcache          pkgs/main/linux-64::fastcache-1.1.0-py37h7b6447c_0\n",
            "  filelock           pkgs/main/noarch::filelock-3.0.12-py_0\n",
            "  flask              pkgs/main/noarch::flask-1.1.1-py_0\n",
            "  fontconfig         pkgs/main/linux-64::fontconfig-2.13.0-h9420a91_0\n",
            "  freetype           pkgs/main/linux-64::freetype-2.9.1-h8a8886c_1\n",
            "  fribidi            pkgs/main/linux-64::fribidi-1.0.5-h7b6447c_0\n",
            "  fsspec             pkgs/main/noarch::fsspec-0.5.2-py_0\n",
            "  future             pkgs/main/linux-64::future-0.17.1-py37_0\n",
            "  get_terminal_size  pkgs/main/linux-64::get_terminal_size-1.0.0-haa9412d_0\n",
            "  gevent             pkgs/main/linux-64::gevent-1.4.0-py37h7b6447c_0\n",
            "  glib               pkgs/main/linux-64::glib-2.56.2-hd408876_0\n",
            "  glob2              pkgs/main/noarch::glob2-0.7-py_0\n",
            "  gmp                pkgs/main/linux-64::gmp-6.1.2-h6c8ec71_1\n",
            "  gmpy2              pkgs/main/linux-64::gmpy2-2.0.8-py37h10f8cd9_2\n",
            "  graphite2          pkgs/main/linux-64::graphite2-1.3.13-h23475e2_0\n",
            "  greenlet           pkgs/main/linux-64::greenlet-0.4.15-py37h7b6447c_0\n",
            "  gst-plugins-base   pkgs/main/linux-64::gst-plugins-base-1.14.0-hbbd80ab_1\n",
            "  gstreamer          pkgs/main/linux-64::gstreamer-1.14.0-hb453b48_1\n",
            "  h5py               pkgs/main/linux-64::h5py-2.9.0-py37h7918eee_0\n",
            "  harfbuzz           pkgs/main/linux-64::harfbuzz-1.8.8-hffaf4a1_0\n",
            "  hdf5               pkgs/main/linux-64::hdf5-1.10.4-hb1b8bf9_0\n",
            "  heapdict           pkgs/main/noarch::heapdict-1.0.1-py_0\n",
            "  html5lib           pkgs/main/linux-64::html5lib-1.0.1-py37_0\n",
            "  icu                pkgs/main/linux-64::icu-58.2-h9c2bf20_1\n",
            "  idna               pkgs/main/linux-64::idna-2.8-py37_0\n",
            "  imageio            pkgs/main/linux-64::imageio-2.6.0-py37_0\n",
            "  imagesize          pkgs/main/linux-64::imagesize-1.1.0-py37_0\n",
            "  importlib_metadata pkgs/main/linux-64::importlib_metadata-0.23-py37_0\n",
            "  intel-openmp       pkgs/main/linux-64::intel-openmp-2019.4-243\n",
            "  ipykernel          pkgs/main/linux-64::ipykernel-5.1.2-py37h39e3cac_0\n",
            "  ipython            pkgs/main/linux-64::ipython-7.8.0-py37h39e3cac_0\n",
            "  ipython_genutils   pkgs/main/linux-64::ipython_genutils-0.2.0-py37_0\n",
            "  ipywidgets         pkgs/main/noarch::ipywidgets-7.5.1-py_0\n",
            "  isort              pkgs/main/linux-64::isort-4.3.21-py37_0\n",
            "  itsdangerous       pkgs/main/linux-64::itsdangerous-1.1.0-py37_0\n",
            "  jbig               pkgs/main/linux-64::jbig-2.1-hdba287a_0\n",
            "  jdcal              pkgs/main/noarch::jdcal-1.4.1-py_0\n",
            "  jedi               pkgs/main/linux-64::jedi-0.15.1-py37_0\n",
            "  jeepney            pkgs/main/noarch::jeepney-0.4.1-py_0\n",
            "  jinja2             pkgs/main/noarch::jinja2-2.10.3-py_0\n",
            "  joblib             pkgs/main/linux-64::joblib-0.13.2-py37_0\n",
            "  jpeg               pkgs/main/linux-64::jpeg-9b-h024ee3a_2\n",
            "  json5              pkgs/main/noarch::json5-0.8.5-py_0\n",
            "  jsonschema         pkgs/main/linux-64::jsonschema-3.0.2-py37_0\n",
            "  jupyter            pkgs/main/linux-64::jupyter-1.0.0-py37_7\n",
            "  jupyter_client     pkgs/main/linux-64::jupyter_client-5.3.3-py37_1\n",
            "  jupyter_console    pkgs/main/linux-64::jupyter_console-6.0.0-py37_0\n",
            "  jupyter_core       pkgs/main/noarch::jupyter_core-4.5.0-py_0\n",
            "  jupyterlab         pkgs/main/noarch::jupyterlab-1.1.4-pyhf63ae98_0\n",
            "  jupyterlab_server  pkgs/main/noarch::jupyterlab_server-1.0.6-py_0\n",
            "  keyring            pkgs/main/linux-64::keyring-18.0.0-py37_0\n",
            "  kiwisolver         pkgs/main/linux-64::kiwisolver-1.1.0-py37he6710b0_0\n",
            "  krb5               pkgs/main/linux-64::krb5-1.16.1-h173b8e3_7\n",
            "  lazy-object-proxy  pkgs/main/linux-64::lazy-object-proxy-1.4.2-py37h7b6447c_0\n",
            "  libarchive         pkgs/main/linux-64::libarchive-3.3.3-h5d8350f_5\n",
            "  libcurl            pkgs/main/linux-64::libcurl-7.65.3-h20c2e04_0\n",
            "  libedit            pkgs/main/linux-64::libedit-3.1.20181209-hc058e9b_0\n",
            "  libffi             pkgs/main/linux-64::libffi-3.2.1-hd88cf55_4\n",
            "  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.1.0-hdf63c60_0\n",
            "  libgfortran-ng     pkgs/main/linux-64::libgfortran-ng-7.3.0-hdf63c60_0\n",
            "  liblief            pkgs/main/linux-64::liblief-0.9.0-h7725739_2\n",
            "  libpng             pkgs/main/linux-64::libpng-1.6.37-hbc83047_0\n",
            "  libsodium          pkgs/main/linux-64::libsodium-1.0.16-h1bed415_0\n",
            "  libssh2            pkgs/main/linux-64::libssh2-1.8.2-h1ba5d50_0\n",
            "  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.1.0-hdf63c60_0\n",
            "  libtiff            pkgs/main/linux-64::libtiff-4.0.10-h2733197_2\n",
            "  libtool            pkgs/main/linux-64::libtool-2.4.6-h7b6447c_5\n",
            "  libuuid            pkgs/main/linux-64::libuuid-1.0.3-h1bed415_2\n",
            "  libxcb             pkgs/main/linux-64::libxcb-1.13-h1bed415_1\n",
            "  libxml2            pkgs/main/linux-64::libxml2-2.9.9-hea5a465_1\n",
            "  libxslt            pkgs/main/linux-64::libxslt-1.1.33-h7d1a2b0_0\n",
            "  llvmlite           pkgs/main/linux-64::llvmlite-0.29.0-py37hd408876_0\n",
            "  locket             pkgs/main/linux-64::locket-0.2.0-py37_1\n",
            "  lxml               pkgs/main/linux-64::lxml-4.4.1-py37hefd8a0e_0\n",
            "  lz4-c              pkgs/main/linux-64::lz4-c-1.8.1.2-h14c3975_0\n",
            "  lzo                pkgs/main/linux-64::lzo-2.10-h49e0be7_2\n",
            "  markupsafe         pkgs/main/linux-64::markupsafe-1.1.1-py37h7b6447c_0\n",
            "  matplotlib         pkgs/main/linux-64::matplotlib-3.1.1-py37h5429711_0\n",
            "  mccabe             pkgs/main/linux-64::mccabe-0.6.1-py37_1\n",
            "  mistune            pkgs/main/linux-64::mistune-0.8.4-py37h7b6447c_0\n",
            "  mkl                pkgs/main/linux-64::mkl-2019.4-243\n",
            "  mkl-service        pkgs/main/linux-64::mkl-service-2.3.0-py37he904b0f_0\n",
            "  mkl_fft            pkgs/main/linux-64::mkl_fft-1.0.14-py37ha843d7b_0\n",
            "  mkl_random         pkgs/main/linux-64::mkl_random-1.1.0-py37hd6b4f25_0\n",
            "  mock               pkgs/main/linux-64::mock-3.0.5-py37_0\n",
            "  more-itertools     pkgs/main/linux-64::more-itertools-7.2.0-py37_0\n",
            "  mpc                pkgs/main/linux-64::mpc-1.1.0-h10f8cd9_1\n",
            "  mpfr               pkgs/main/linux-64::mpfr-4.0.1-hdf1c602_3\n",
            "  mpmath             pkgs/main/linux-64::mpmath-1.1.0-py37_0\n",
            "  msgpack-python     pkgs/main/linux-64::msgpack-python-0.6.1-py37hfd86e86_1\n",
            "  multipledispatch   pkgs/main/linux-64::multipledispatch-0.6.0-py37_0\n",
            "  navigator-updater  pkgs/main/linux-64::navigator-updater-0.2.1-py37_0\n",
            "  nbconvert          pkgs/main/linux-64::nbconvert-5.6.0-py37_1\n",
            "  nbformat           pkgs/main/linux-64::nbformat-4.4.0-py37_0\n",
            "  ncurses            pkgs/main/linux-64::ncurses-6.1-he6710b0_1\n",
            "  networkx           pkgs/main/noarch::networkx-2.3-py_0\n",
            "  nltk               pkgs/main/linux-64::nltk-3.4.5-py37_0\n",
            "  nose               pkgs/main/linux-64::nose-1.3.7-py37_2\n",
            "  notebook           pkgs/main/linux-64::notebook-6.0.1-py37_0\n",
            "  numba              pkgs/main/linux-64::numba-0.45.1-py37h962f231_0\n",
            "  numexpr            pkgs/main/linux-64::numexpr-2.7.0-py37h9e4a6bb_0\n",
            "  numpy              pkgs/main/linux-64::numpy-1.17.2-py37haad9e8e_0\n",
            "  numpy-base         pkgs/main/linux-64::numpy-base-1.17.2-py37hde5b4d6_0\n",
            "  numpydoc           pkgs/main/noarch::numpydoc-0.9.1-py_0\n",
            "  olefile            pkgs/main/linux-64::olefile-0.46-py37_0\n",
            "  openpyxl           pkgs/main/noarch::openpyxl-3.0.0-py_0\n",
            "  openssl            pkgs/main/linux-64::openssl-1.1.1d-h7b6447c_2\n",
            "  packaging          pkgs/main/noarch::packaging-19.2-py_0\n",
            "  pandas             pkgs/main/linux-64::pandas-0.25.1-py37he6710b0_0\n",
            "  pandoc             pkgs/main/linux-64::pandoc-2.2.3.2-0\n",
            "  pandocfilters      pkgs/main/linux-64::pandocfilters-1.4.2-py37_1\n",
            "  pango              pkgs/main/linux-64::pango-1.42.4-h049681c_0\n",
            "  parso              pkgs/main/noarch::parso-0.5.1-py_0\n",
            "  partd              pkgs/main/noarch::partd-1.0.0-py_0\n",
            "  patchelf           pkgs/main/linux-64::patchelf-0.9-he6710b0_3\n",
            "  path.py            pkgs/main/noarch::path.py-12.0.1-py_0\n",
            "  pathlib2           pkgs/main/linux-64::pathlib2-2.3.5-py37_0\n",
            "  patsy              pkgs/main/linux-64::patsy-0.5.1-py37_0\n",
            "  pcre               pkgs/main/linux-64::pcre-8.43-he6710b0_0\n",
            "  pep8               pkgs/main/linux-64::pep8-1.7.1-py37_0\n",
            "  pexpect            pkgs/main/linux-64::pexpect-4.7.0-py37_0\n",
            "  pickleshare        pkgs/main/linux-64::pickleshare-0.7.5-py37_0\n",
            "  pillow             pkgs/main/linux-64::pillow-6.2.0-py37h34e0f95_0\n",
            "  pip                pkgs/main/linux-64::pip-19.2.3-py37_0\n",
            "  pixman             pkgs/main/linux-64::pixman-0.38.0-h7b6447c_0\n",
            "  pkginfo            pkgs/main/linux-64::pkginfo-1.5.0.1-py37_0\n",
            "  pluggy             pkgs/main/linux-64::pluggy-0.13.0-py37_0\n",
            "  ply                pkgs/main/linux-64::ply-3.11-py37_0\n",
            "  prometheus_client  pkgs/main/noarch::prometheus_client-0.7.1-py_0\n",
            "  prompt_toolkit     pkgs/main/noarch::prompt_toolkit-2.0.10-py_0\n",
            "  psutil             pkgs/main/linux-64::psutil-5.6.3-py37h7b6447c_0\n",
            "  ptyprocess         pkgs/main/linux-64::ptyprocess-0.6.0-py37_0\n",
            "  py                 pkgs/main/linux-64::py-1.8.0-py37_0\n",
            "  py-lief            pkgs/main/linux-64::py-lief-0.9.0-py37h7725739_2\n",
            "  pycodestyle        pkgs/main/linux-64::pycodestyle-2.5.0-py37_0\n",
            "  pycosat            pkgs/main/linux-64::pycosat-0.6.3-py37h14c3975_0\n",
            "  pycparser          pkgs/main/linux-64::pycparser-2.19-py37_0\n",
            "  pycrypto           pkgs/main/linux-64::pycrypto-2.6.1-py37h14c3975_9\n",
            "  pycurl             pkgs/main/linux-64::pycurl-7.43.0.3-py37h1ba5d50_0\n",
            "  pyflakes           pkgs/main/linux-64::pyflakes-2.1.1-py37_0\n",
            "  pygments           pkgs/main/noarch::pygments-2.4.2-py_0\n",
            "  pylint             pkgs/main/linux-64::pylint-2.4.2-py37_0\n",
            "  pyodbc             pkgs/main/linux-64::pyodbc-4.0.27-py37he6710b0_0\n",
            "  pyopenssl          pkgs/main/linux-64::pyopenssl-19.0.0-py37_0\n",
            "  pyparsing          pkgs/main/noarch::pyparsing-2.4.2-py_0\n",
            "  pyqt               pkgs/main/linux-64::pyqt-5.9.2-py37h05f1152_2\n",
            "  pyrsistent         pkgs/main/linux-64::pyrsistent-0.15.4-py37h7b6447c_0\n",
            "  pysocks            pkgs/main/linux-64::pysocks-1.7.1-py37_0\n",
            "  pytables           pkgs/main/linux-64::pytables-3.5.2-py37h71ec239_1\n",
            "  pytest             pkgs/main/linux-64::pytest-5.2.1-py37_0\n",
            "  pytest-arraydiff   pkgs/main/linux-64::pytest-arraydiff-0.3-py37h39e3cac_0\n",
            "  pytest-astropy     pkgs/main/linux-64::pytest-astropy-0.5.0-py37_0\n",
            "  pytest-doctestplus pkgs/main/noarch::pytest-doctestplus-0.4.0-py_0\n",
            "  pytest-openfiles   pkgs/main/noarch::pytest-openfiles-0.4.0-py_0\n",
            "  pytest-remotedata  pkgs/main/linux-64::pytest-remotedata-0.3.2-py37_0\n",
            "  python             pkgs/main/linux-64::python-3.7.4-h265db76_1\n",
            "  python-dateutil    pkgs/main/linux-64::python-dateutil-2.8.0-py37_0\n",
            "  python-libarchive~ pkgs/main/linux-64::python-libarchive-c-2.8-py37_13\n",
            "  pytz               pkgs/main/noarch::pytz-2019.3-py_0\n",
            "  pywavelets         pkgs/main/linux-64::pywavelets-1.0.3-py37hdd07704_1\n",
            "  pyyaml             pkgs/main/linux-64::pyyaml-5.1.2-py37h7b6447c_0\n",
            "  pyzmq              pkgs/main/linux-64::pyzmq-18.1.0-py37he6710b0_0\n",
            "  qt                 pkgs/main/linux-64::qt-5.9.7-h5867ecd_1\n",
            "  qtawesome          pkgs/main/noarch::qtawesome-0.6.0-py_0\n",
            "  qtconsole          pkgs/main/noarch::qtconsole-4.5.5-py_0\n",
            "  qtpy               pkgs/main/noarch::qtpy-1.9.0-py_0\n",
            "  readline           pkgs/main/linux-64::readline-7.0-h7b6447c_5\n",
            "  requests           pkgs/main/linux-64::requests-2.22.0-py37_0\n",
            "  ripgrep            pkgs/main/linux-64::ripgrep-0.10.0-hc07d326_0\n",
            "  rope               pkgs/main/noarch::rope-0.14.0-py_0\n",
            "  ruamel_yaml        pkgs/main/linux-64::ruamel_yaml-0.15.46-py37h14c3975_0\n",
            "  scikit-image       pkgs/main/linux-64::scikit-image-0.15.0-py37he6710b0_0\n",
            "  scikit-learn       pkgs/main/linux-64::scikit-learn-0.21.3-py37hd81dba3_0\n",
            "  scipy              pkgs/main/linux-64::scipy-1.3.1-py37h7c811a0_0\n",
            "  seaborn            pkgs/main/linux-64::seaborn-0.9.0-py37_0\n",
            "  secretstorage      pkgs/main/linux-64::secretstorage-3.1.1-py37_0\n",
            "  send2trash         pkgs/main/linux-64::send2trash-1.5.0-py37_0\n",
            "  setuptools         pkgs/main/linux-64::setuptools-41.4.0-py37_0\n",
            "  simplegeneric      pkgs/main/linux-64::simplegeneric-0.8.1-py37_2\n",
            "  singledispatch     pkgs/main/linux-64::singledispatch-3.4.0.3-py37_0\n",
            "  sip                pkgs/main/linux-64::sip-4.19.8-py37hf484d3e_0\n",
            "  six                pkgs/main/linux-64::six-1.12.0-py37_0\n",
            "  snappy             pkgs/main/linux-64::snappy-1.1.7-hbae5bb6_3\n",
            "  snowballstemmer    pkgs/main/noarch::snowballstemmer-2.0.0-py_0\n",
            "  sortedcollections  pkgs/main/linux-64::sortedcollections-1.1.2-py37_0\n",
            "  sortedcontainers   pkgs/main/linux-64::sortedcontainers-2.1.0-py37_0\n",
            "  soupsieve          pkgs/main/linux-64::soupsieve-1.9.3-py37_0\n",
            "  sphinx             pkgs/main/noarch::sphinx-2.2.0-py_0\n",
            "  sphinxcontrib      pkgs/main/linux-64::sphinxcontrib-1.0-py37_1\n",
            "  sphinxcontrib-app~ pkgs/main/noarch::sphinxcontrib-applehelp-1.0.1-py_0\n",
            "  sphinxcontrib-dev~ pkgs/main/noarch::sphinxcontrib-devhelp-1.0.1-py_0\n",
            "  sphinxcontrib-htm~ pkgs/main/noarch::sphinxcontrib-htmlhelp-1.0.2-py_0\n",
            "  sphinxcontrib-jsm~ pkgs/main/noarch::sphinxcontrib-jsmath-1.0.1-py_0\n",
            "  sphinxcontrib-qth~ pkgs/main/noarch::sphinxcontrib-qthelp-1.0.2-py_0\n",
            "  sphinxcontrib-ser~ pkgs/main/noarch::sphinxcontrib-serializinghtml-1.1.3-py_0\n",
            "  sphinxcontrib-web~ pkgs/main/noarch::sphinxcontrib-websupport-1.1.2-py_0\n",
            "  spyder             pkgs/main/linux-64::spyder-3.3.6-py37_0\n",
            "  spyder-kernels     pkgs/main/linux-64::spyder-kernels-0.5.2-py37_0\n",
            "  sqlalchemy         pkgs/main/linux-64::sqlalchemy-1.3.9-py37h7b6447c_0\n",
            "  sqlite             pkgs/main/linux-64::sqlite-3.30.0-h7b6447c_0\n",
            "  statsmodels        pkgs/main/linux-64::statsmodels-0.10.1-py37hdd07704_0\n",
            "  sympy              pkgs/main/linux-64::sympy-1.4-py37_0\n",
            "  tbb                pkgs/main/linux-64::tbb-2019.4-hfd86e86_0\n",
            "  tblib              pkgs/main/noarch::tblib-1.4.0-py_0\n",
            "  terminado          pkgs/main/linux-64::terminado-0.8.2-py37_0\n",
            "  testpath           pkgs/main/linux-64::testpath-0.4.2-py37_0\n",
            "  tk                 pkgs/main/linux-64::tk-8.6.8-hbc83047_0\n",
            "  toolz              pkgs/main/noarch::toolz-0.10.0-py_0\n",
            "  tornado            pkgs/main/linux-64::tornado-6.0.3-py37h7b6447c_0\n",
            "  tqdm               pkgs/main/noarch::tqdm-4.36.1-py_0\n",
            "  traitlets          pkgs/main/linux-64::traitlets-4.3.3-py37_0\n",
            "  unicodecsv         pkgs/main/linux-64::unicodecsv-0.14.1-py37_0\n",
            "  unixodbc           pkgs/main/linux-64::unixodbc-2.3.7-h14c3975_0\n",
            "  urllib3            pkgs/main/linux-64::urllib3-1.24.2-py37_0\n",
            "  wcwidth            pkgs/main/linux-64::wcwidth-0.1.7-py37_0\n",
            "  webencodings       pkgs/main/linux-64::webencodings-0.5.1-py37_1\n",
            "  werkzeug           pkgs/main/noarch::werkzeug-0.16.0-py_0\n",
            "  wheel              pkgs/main/linux-64::wheel-0.33.6-py37_0\n",
            "  widgetsnbextension pkgs/main/linux-64::widgetsnbextension-3.5.1-py37_0\n",
            "  wrapt              pkgs/main/linux-64::wrapt-1.11.2-py37h7b6447c_0\n",
            "  wurlitzer          pkgs/main/linux-64::wurlitzer-1.0.3-py37_0\n",
            "  xlrd               pkgs/main/linux-64::xlrd-1.2.0-py37_0\n",
            "  xlsxwriter         pkgs/main/noarch::xlsxwriter-1.2.1-py_0\n",
            "  xlwt               pkgs/main/linux-64::xlwt-1.3.0-py37_0\n",
            "  xz                 pkgs/main/linux-64::xz-5.2.4-h14c3975_4\n",
            "  yaml               pkgs/main/linux-64::yaml-0.1.7-had09818_2\n",
            "  zeromq             pkgs/main/linux-64::zeromq-4.3.1-he6710b0_3\n",
            "  zict               pkgs/main/noarch::zict-1.0.0-py_0\n",
            "  zipp               pkgs/main/noarch::zipp-0.6.0-py_0\n",
            "  zlib               pkgs/main/linux-64::zlib-1.2.11-h7b6447c_3\n",
            "  zstd               pkgs/main/linux-64::zstd-1.3.7-h0b5b093_0\n",
            "\n",
            "\n",
            "Preparing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Executing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "installation finished.\n",
            "WARNING:\n",
            "    You currently have a PYTHONPATH environment variable set. This may cause\n",
            "    unexpected behavior when running the Python interpreter in Anaconda3.\n",
            "    For best results, please verify that your PYTHONPATH only points to\n",
            "    directories of packages that are compatible with the Python interpreter\n",
            "    in Anaconda3: /usr/local\n",
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Solving environment: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bfailed with initial frozen solve. Retrying with flexible solve.\n",
            "Solving environment: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bfailed with repodata from current_repodata.json, will retry with next repodata source.\n",
            "Collecting package metadata (repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "Solving environment: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.7.12\n",
            "  latest version: 4.10.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /usr/local\n",
            "\n",
            "  added / updated specs:\n",
            "    - deepchem-gpu=2.3.0\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    _py-xgboost-mutex-2.0      |            cpu_0           8 KB  conda-forge\n",
            "    _tflow_select-2.1.0        |              gpu           2 KB\n",
            "    absl-py-0.12.0             |     pyhd8ed1ab_0          96 KB  conda-forge\n",
            "    astor-0.8.1                |     pyh9f0ad1d_0          25 KB  conda-forge\n",
            "    astunparse-1.6.3           |     pyhd8ed1ab_0          15 KB  conda-forge\n",
            "    c-ares-1.17.1              |       h36c2ea0_0         111 KB  conda-forge\n",
            "    certifi-2019.9.11          |           py37_0         147 KB  conda-forge\n",
            "    conda-4.10.1               |   py37h89c1867_0         3.1 MB  conda-forge\n",
            "    cudatoolkit-10.1.243       |       h6bb024c_0       347.4 MB\n",
            "    cudnn-7.6.5.32             |       hc0a50b0_1       250.7 MB  conda-forge\n",
            "    cupti-10.1.168             |                0         1.4 MB\n",
            "    deepchem-gpu-2.3.0         |           py37_0         2.1 MB  deepchem\n",
            "    fftw3f-3.3.4               |                2         1.2 MB  omnia\n",
            "    gast-0.4.0                 |     pyh9f0ad1d_0          12 KB  conda-forge\n",
            "    google-pasta-0.2.0         |     pyh8c360ce_0          42 KB  conda-forge\n",
            "    grpcio-1.23.0              |   py37hb0870dc_1         1.1 MB  conda-forge\n",
            "    importlib-metadata-3.10.1  |   py37h89c1867_0          27 KB  conda-forge\n",
            "    keras-applications-1.0.8   |             py_1          30 KB  conda-forge\n",
            "    keras-preprocessing-1.1.2  |     pyhd8ed1ab_0          34 KB  conda-forge\n",
            "    libboost-1.67.0            |       h46d08c1_4        13.0 MB\n",
            "    libprotobuf-3.13.0.1       |       h8b12597_0         2.3 MB  conda-forge\n",
            "    libxgboost-1.2.0           |       he1b5a44_0         3.1 MB  conda-forge\n",
            "    markdown-3.3.4             |     pyhd8ed1ab_0          67 KB  conda-forge\n",
            "    mdtraj-1.9.5               |   py37h113d463_0         1.7 MB  conda-forge\n",
            "    openmm-7.4.2               |py37_cuda101_rc_1        11.9 MB  omnia\n",
            "    pdbfixer-1.6               |             py_1         167 KB  omnia\n",
            "    protobuf-3.13.0.1          |   py37h745909e_1         704 KB  conda-forge\n",
            "    py-boost-1.67.0            |   py37h04863e7_4         278 KB\n",
            "    py-xgboost-1.2.0           |   py37hc8dfbb8_0         1.7 MB  conda-forge\n",
            "    python_abi-3.7             |          1_cp37m           4 KB  conda-forge\n",
            "    rdkit-2020.03.3.0          |   py37hc20afe1_1        24.8 MB  rdkit\n",
            "    simdna-0.4.2               |             py_0         627 KB  deepchem\n",
            "    tensorboard-1.14.0         |           py37_0         3.2 MB  conda-forge\n",
            "    tensorflow-1.14.0          |gpu_py37h74c33d7_0           4 KB\n",
            "    tensorflow-base-1.14.0     |gpu_py37he45bfe2_0       146.3 MB\n",
            "    tensorflow-estimator-1.14.0|   py37h5ca1d4c_0         645 KB  conda-forge\n",
            "    tensorflow-gpu-1.14.0      |       h0d30ee6_0           3 KB\n",
            "    termcolor-1.1.0            |             py_2           6 KB  conda-forge\n",
            "    typing_extensions-3.7.4.3  |             py_0          25 KB  conda-forge\n",
            "    xgboost-1.2.0              |   py37h3340039_0          11 KB  conda-forge\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:       817.9 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _py-xgboost-mutex  conda-forge/linux-64::_py-xgboost-mutex-2.0-cpu_0\n",
            "  _tflow_select      pkgs/main/linux-64::_tflow_select-2.1.0-gpu\n",
            "  absl-py            conda-forge/noarch::absl-py-0.12.0-pyhd8ed1ab_0\n",
            "  astor              conda-forge/noarch::astor-0.8.1-pyh9f0ad1d_0\n",
            "  astunparse         conda-forge/noarch::astunparse-1.6.3-pyhd8ed1ab_0\n",
            "  c-ares             conda-forge/linux-64::c-ares-1.17.1-h36c2ea0_0\n",
            "  cudatoolkit        pkgs/main/linux-64::cudatoolkit-10.1.243-h6bb024c_0\n",
            "  cudnn              conda-forge/linux-64::cudnn-7.6.5.32-hc0a50b0_1\n",
            "  cupti              pkgs/main/linux-64::cupti-10.1.168-0\n",
            "  deepchem-gpu       deepchem/linux-64::deepchem-gpu-2.3.0-py37_0\n",
            "  fftw3f             omnia/linux-64::fftw3f-3.3.4-2\n",
            "  gast               conda-forge/noarch::gast-0.4.0-pyh9f0ad1d_0\n",
            "  google-pasta       conda-forge/noarch::google-pasta-0.2.0-pyh8c360ce_0\n",
            "  grpcio             conda-forge/linux-64::grpcio-1.23.0-py37hb0870dc_1\n",
            "  importlib-metadata conda-forge/linux-64::importlib-metadata-3.10.1-py37h89c1867_0\n",
            "  keras-applications conda-forge/noarch::keras-applications-1.0.8-py_1\n",
            "  keras-preprocessi~ conda-forge/noarch::keras-preprocessing-1.1.2-pyhd8ed1ab_0\n",
            "  libboost           pkgs/main/linux-64::libboost-1.67.0-h46d08c1_4\n",
            "  libprotobuf        conda-forge/linux-64::libprotobuf-3.13.0.1-h8b12597_0\n",
            "  libxgboost         conda-forge/linux-64::libxgboost-1.2.0-he1b5a44_0\n",
            "  markdown           conda-forge/noarch::markdown-3.3.4-pyhd8ed1ab_0\n",
            "  mdtraj             conda-forge/linux-64::mdtraj-1.9.5-py37h113d463_0\n",
            "  openmm             omnia/linux-64::openmm-7.4.2-py37_cuda101_rc_1\n",
            "  pdbfixer           omnia/noarch::pdbfixer-1.6-py_1\n",
            "  protobuf           conda-forge/linux-64::protobuf-3.13.0.1-py37h745909e_1\n",
            "  py-boost           pkgs/main/linux-64::py-boost-1.67.0-py37h04863e7_4\n",
            "  py-xgboost         conda-forge/linux-64::py-xgboost-1.2.0-py37hc8dfbb8_0\n",
            "  python_abi         conda-forge/linux-64::python_abi-3.7-1_cp37m\n",
            "  rdkit              rdkit/linux-64::rdkit-2020.03.3.0-py37hc20afe1_1\n",
            "  simdna             deepchem/noarch::simdna-0.4.2-py_0\n",
            "  tensorboard        conda-forge/linux-64::tensorboard-1.14.0-py37_0\n",
            "  tensorflow         pkgs/main/linux-64::tensorflow-1.14.0-gpu_py37h74c33d7_0\n",
            "  tensorflow-base    pkgs/main/linux-64::tensorflow-base-1.14.0-gpu_py37he45bfe2_0\n",
            "  tensorflow-estima~ conda-forge/linux-64::tensorflow-estimator-1.14.0-py37h5ca1d4c_0\n",
            "  tensorflow-gpu     pkgs/main/linux-64::tensorflow-gpu-1.14.0-h0d30ee6_0\n",
            "  termcolor          conda-forge/noarch::termcolor-1.1.0-py_2\n",
            "  typing_extensions  conda-forge/noarch::typing_extensions-3.7.4.3-py_0\n",
            "  xgboost            conda-forge/linux-64::xgboost-1.2.0-py37h3340039_0\n",
            "\n",
            "The following packages will be UPDATED:\n",
            "\n",
            "  conda                      pkgs/main::conda-4.7.12-py37_0 --> conda-forge::conda-4.10.1-py37h89c1867_0\n",
            "\n",
            "The following packages will be SUPERSEDED by a higher-priority channel:\n",
            "\n",
            "  certifi                                         pkgs/main --> conda-forge\n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages\n",
            "typing_extensions-3. | 25 KB     | : 100% 1.0/1 [00:00<00:00,  8.86it/s]               \n",
            "libprotobuf-3.13.0.1 | 2.3 MB    | : 100% 1.0/1 [00:00<00:00,  2.00it/s]              \n",
            "py-boost-1.67.0      | 278 KB    | : 100% 1.0/1 [00:00<00:00,  3.73it/s]               \n",
            "mdtraj-1.9.5         | 1.7 MB    | : 100% 1.0/1 [00:00<00:00,  2.19it/s]\n",
            "cudatoolkit-10.1.243 | 347.4 MB  | : 100% 1.0/1 [00:09<00:00,  9.11s/it]               \n",
            "simdna-0.4.2         | 627 KB    | : 100% 1.0/1 [00:00<00:00,  3.13it/s]\n",
            "google-pasta-0.2.0   | 42 KB     | : 100% 1.0/1 [00:00<00:00, 14.88it/s]\n",
            "cupti-10.1.168       | 1.4 MB    | : 100% 1.0/1 [00:00<00:00,  5.68it/s]\n",
            "tensorflow-base-1.14 | 146.3 MB  | : 100% 1.0/1 [00:27<00:00, 27.98s/it]               \n",
            "certifi-2019.9.11    | 147 KB    | : 100% 1.0/1 [00:00<00:00,  6.64it/s]\n",
            "tensorboard-1.14.0   | 3.2 MB    | : 100% 1.0/1 [00:00<00:00,  1.89it/s]               \n",
            "py-xgboost-1.2.0     | 1.7 MB    | : 100% 1.0/1 [00:00<00:00,  2.23it/s]\n",
            "xgboost-1.2.0        | 11 KB     | : 100% 1.0/1 [00:00<00:00, 21.71it/s]\n",
            "cudnn-7.6.5.32       | 250.7 MB  | : 100% 1.0/1 [00:43<00:00, 43.94s/it]               \n",
            "_tflow_select-2.1.0  | 2 KB      | : 100% 1.0/1 [00:00<00:00,  7.90it/s]\n",
            "protobuf-3.13.0.1    | 704 KB    | : 100% 1.0/1 [00:00<00:00,  2.93it/s]\n",
            "openmm-7.4.2         | 11.9 MB   | : 100% 1.0/1 [00:03<00:00,  3.57s/it]\n",
            "deepchem-gpu-2.3.0   | 2.1 MB    | : 100% 1.0/1 [00:00<00:00,  1.40it/s]\n",
            "termcolor-1.1.0      | 6 KB      | : 100% 1.0/1 [00:00<00:00, 25.19it/s]\n",
            "keras-preprocessing- | 34 KB     | : 100% 1.0/1 [00:00<00:00, 18.46it/s]\n",
            "fftw3f-3.3.4         | 1.2 MB    | : 100% 1.0/1 [00:00<00:00,  2.58it/s]\n",
            "tensorflow-1.14.0    | 4 KB      | : 100% 1.0/1 [00:00<00:00,  8.44it/s]\n",
            "markdown-3.3.4       | 67 KB     | : 100% 1.0/1 [00:00<00:00, 11.16it/s]\n",
            "keras-applications-1 | 30 KB     | : 100% 1.0/1 [00:00<00:00, 13.89it/s]\n",
            "libboost-1.67.0      | 13.0 MB   | : 100% 1.0/1 [00:02<00:00,  1.44s/it]               \n",
            "grpcio-1.23.0        | 1.1 MB    | : 100% 1.0/1 [00:00<00:00,  3.11it/s]\n",
            "conda-4.10.1         | 3.1 MB    | : 100% 1.0/1 [00:00<00:00,  1.21it/s]\n",
            "importlib-metadata-3 | 27 KB     | : 100% 1.0/1 [00:00<00:00, 20.65it/s]\n",
            "astunparse-1.6.3     | 15 KB     | : 100% 1.0/1 [00:00<00:00, 20.86it/s]\n",
            "tensorflow-gpu-1.14. | 3 KB      | : 100% 1.0/1 [00:00<00:00,  8.21it/s]\n",
            "libxgboost-1.2.0     | 3.1 MB    | : 100% 1.0/1 [00:00<00:00,  1.42it/s]\n",
            "absl-py-0.12.0       | 96 KB     | : 100% 1.0/1 [00:00<00:00, 14.93it/s]\n",
            "gast-0.4.0           | 12 KB     | : 100% 1.0/1 [00:00<00:00, 21.10it/s]\n",
            "astor-0.8.1          | 25 KB     | : 100% 1.0/1 [00:00<00:00, 17.16it/s]\n",
            "rdkit-2020.03.3.0    | 24.8 MB   | : 100% 1.0/1 [00:05<00:00,  5.28s/it]\n",
            "pdbfixer-1.6         | 167 KB    | : 100% 1.0/1 [00:00<00:00,  8.67it/s]\n",
            "tensorflow-estimator | 645 KB    | : 100% 1.0/1 [00:00<00:00,  3.08it/s]\n",
            "python_abi-3.7       | 4 KB      | : 100% 1.0/1 [00:00<00:00, 20.36it/s]\n",
            "_py-xgboost-mutex-2. | 8 KB      | : 100% 1.0/1 [00:00<00:00, 16.35it/s]\n",
            "c-ares-1.17.1        | 111 KB    | : 100% 1.0/1 [00:00<00:00, 12.96it/s]\n",
            "Preparing transaction: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "Verifying transaction: / \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Executing transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ b'By downloading and using the cuDNN conda packages, you accept the terms and conditions of the NVIDIA cuDNN EULA -\\n  https://docs.nvidia.com/deeplearning/cudnn/sla/index.html\\n'\n",
            "\b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/sklearn/externals/joblib/__init__.py:15: FutureWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.\n",
            "  warnings.warn(msg, category=FutureWarning)\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "stream",
          "text": [
            "WARNING:tensorflow:\n",
            "The TensorFlow contrib module will not be included in TensorFlow 2.0.\n",
            "For more information, please see:\n",
            "  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md\n",
            "  * https://github.com/tensorflow/addons\n",
            "  * https://github.com/tensorflow/io (for I/O related ops)\n",
            "If you depend on functionality not listed there, please file an issue.\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ygWxSzB9SFXm"
      },
      "source": [
        "Developing new medicine is a very time-consuming, labor-intensive, and expensive task which begins with rounds of screening process where researchers run some assays with thousands of molecules to identify potential drug candidates to go onto a clinical trial. The field of cheminformatics have been developed to help accelerate this process by performing laboratory experiments on computers. \n",
        "\n",
        "In order to this we need to find a way to represent a molecule so that we can train machine learning models on it.  As mentioned above, this is a critical step.   If we were doing physics we would want to represent the molecule in the most natural way to represent the physics. For example there are features of the molecules that *cause* it to bind to the protein in question -- there are specific interactions between the chemical groups of the molecule and the amino acids. From a physical perspective, identifying these features is the most important part of the problem and we would like the representation to emphasize these.\n",
        "\n",
        "But given that we don't have any idea what is happening, we need another approach. We somehow have to represent both molecules and proteins in a way so that we can train machine learning models on them to see if it is possible to associate them with each other.\n",
        "\n",
        "For small molecules, this is a classic problem. How do we take a molecule and represent it. There is a field that has addressed this called 'cheminformatics'.  \n",
        "\n",
        "A particularly simple method is for molecules are represented as text strings called SMILES (“Simplified Molecular-Input Line-Entry System”). Please refer to [this page](https://www.daylight.com/dayhtml/doc/theory/theory.smiles.html) for more explanantion. Here is an example SMILES string and the molecule it represents, which we obtain via `MolFromSmiles` function of Rdkit."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WobQ8Nz4U-ke",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "8b4f4bc4-2fec-4f08-ecf8-9f3dfcf83518"
      },
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive')\n",
        "G_PATH = './drive/MyDrive/AM216/Drug Binding/'"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Mounted at /content/drive\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IcdYdzPuHNfH"
      },
      "source": [
        "# %%capture\n",
        "# !unzip data_Drug_target_binding_affinity.zip"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "M8YkkNgGSFXo"
      },
      "source": [
        "import numpy as np\n",
        "import rdkit\n",
        "from rdkit.Chem import Draw\n",
        "from rdkit.Chem import AllChem as Chem\n",
        "from rdkit.Chem import MolFromSmiles"
      ],
      "execution_count": 5,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "b4sRfDxkSFXs",
        "scrolled": false,
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 385
        },
        "outputId": "53c25a09-df8a-4ff1-e199-456e1b6a5ee3"
      },
      "source": [
        "smiles = []\n",
        "with open(G_PATH + 'data_Drug_target_binding_affinity/Sample_Ligand_List.txt', 'r') as f:\n",
        "    for line in f:\n",
        "        smiles.append(line[:-1])\n",
        "        \n",
        "print('SMILES string:')\n",
        "print(smiles[0])\n",
        "\n",
        "print('\\nOriginal molecule:')\n",
        "mols = MolFromSmiles(smiles[0])\n",
        "Draw.MolToImage(mols)"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "SMILES string:\n",
            "O[C@@H](CNC1CCN(CC1)c2ccc(CC3SC(=O)NC3=O)cc2)c4ccc(O)c(NS(=O)(=O)c5ccccc5)c4\n",
            "\n",
            "Original molecule:\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAASwAAAEsCAIAAAD2HxkiAAAkLUlEQVR4nO3deVhV1d4H8N85BxAZhBRNRUspJ8QpxAkbUEpF1Aot04t6tce0DOm1MoekQXMio7xO3ZxuotcBFXDMgRwQgyMqkhNeQBSQQZmnM633j4UnRNQzbFiQ38/T0/MI7LXXgfM9e+81yhhjBADiyEVXAOBphxACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAgiGEAIIhhACCIYQAghWX0P41VdERPPn0759tGwZhYSIrQ5A7amvIdSLiaHPP6esLNH1AKgt9TWE5eUUEkIpKaTTERHJ62s9AcxWX9/c1tYUGEjt29OAARQcTM2b09ixtGGD6GoBSE/GGBNdBwPs2kVjxpClJYWH07BhomsDIKX6eiWsZvRomjuX1GoaPZpiYkTXBkBKDeRKSESM0fvv04YN5OREp09Tp06iKwQgjYYTQiLSasnPj8LDqX17OnOGWrYUXSEACTSQ21FOoaDQUPLwYAUF3334YVFRkegKAUigQYWQiGxtaf/+j3r0mLdnj5+fn0qlMuporVabkJBQXl5eS7UDMEGDuh29LyUlZcCAAXfu3Bk7dmxoaKj8sb2IGRkZ5+47ffp0fn7+pEmTunTp8vnnn9dZhQEeo0GGkIguXbr08ssvFxQUfPbZZ8uWLav6rdzc3NjY2Li4OP7/nJycqt91dnbOyMggos2bN/v7+9dppQFq0lBDSERRUVHDhg2rqKhYvny5l5fX6dOn+eXuypUrVV+Ug4ODm5vbwIEDPT09+/bt26JFizVr1nz44YeWlpYRERFDhw4V+BIAqEGHkIi2bNkyYcIEW1vb4uJi/RdtbW179uzpfp+rq2tpaWl8fHxsbGyvXr0GDRpERLNnz162bJmNjc2xY8f69esn7hUANPAQElHHjh3T0tKee+45Ly+vPn36eHh4dO3alTF27do1/aNgXFwcb8KZOnXqunXriIgxNnny5E2bNjk5OUVHR3fs2FH064CnV8MO4Z07d5ydna2trbOzs21tbY8ePbp///64uLj4+PiysjL9j1lYWHTr1q1Pnz4+Pj4jR47kX1Sr1SNHjjx06JCLi8uZM2eeffZZQS8CnnYWoitglp07d+p0umHDhtna2hLRoUOHQu7PPGzVqpW7uzt/FHzppZdsbGyqHWtpaRkWFjZ48OCzZ8/6+vpGRUXZ2dnVcf0BqKGHMCwsjIj8/Pz4P998801HR0cPDw8PD4+mTZs+8XAbG5uIiIiBAwcqlcpRo0YdPHjQysqqdmsM8JAGfDualZXl7OxsYWGRnZ3dpEkTk8tJTk4eMGBAVlbWuHHjtmzZIpPJJKwkwBM1tBEzVezevVur1Q4ZMsScBBKRi4vLvn377Ozstm7dOnfuXKmqB2CgBhzCavei5ujdu3d4eLiVldWSJUt+/PFH8wsEMFxDvR29e/duy5YtZTJZVlbWM888I0mZW7du/cc//iGTyf773/+OGTNGkjIBnqihXgn37Nmj0Whef/11qRJIROPGjVu4cKFMJuPj2gDqRkMNoYT3okSUnJzM7whee+01rVYbghUWoQ41yBDm5+cfP37cwsJC3/NuDpVK5e7u3q5du/z8/F27dhHR6NGjzS8WwEANMoR79+5VqVReXl5OTk7ml3bkyJH8/PymTZs6ODjs3r2bpLvAwsMwmfNhDbKzXtp7UX1pcXFxN2/ebNOmTd++fSUpGYiosLAwISHh3Llz0dHRp06dev3111u3br1o0SKFQiG6avVFwwthUVHR0aNHFQrFW2+9ZX5pGo0mMjKSiPz8/DZt2kREo0ePRn+9OcrKys6fP6+f0nnjxo2q392zZ09xcXFRUdGqVatE1bC+aUghLCkpiY+PX7t2bXl5+cCBA1u0aEFEN2/evHz58jBTFyM9fvx4bm6um5tbly5dpL3APj20Wu3Vq1f1c1aUSmVFRYX+u5aWlt27d/f09OQzywoLC729vVevXu3s7IyhEZVYPaZWs/j48+vWrZsyZUr37t31NzBt2rR5/vnn8/LyMjMzW7dubW1tffLkSdNOMXXqVCIKCgo6d+4cEbVs2VKj0Uj7Kv7GNBrN77//zkfP6ykUiu7du0+ZMmXdunXnz59Xq9VqtfrixYuRkZH8qPDwcIVCIZPJ1q9fL7b+9US9C2F6OouIYEFBzNub2dgwF5fh+r+uhYVFr169Jk6c2KZNGyLy9vauqKgICAggIgcHhwsXLhh7Lo1Gwy+nCQkJ/FP5o48+qo0X9Xe1d+9e/gts1aqVr69vUFBQREREXl4eYyw9PT0iIiIoKMjb25un1N7eXqvV8gPXrl3L47p3716RL6B+EBbCoCDGGJs3jwUGspgY9t13bOhQ1qwZI3rgv7feCh03blxISEh0dHRpaSk/Njk5uWXLlkQ0duxYtVrNexScnZ1v3rxpVB2OHz9ORB07dmSMde7cmYiOHz8u8ev8Wxs/fjy/j2CMFRUV7d+/PygoyMfH5+FWaxcXl7Fjx/J8cvxTz8bG5syZM6LqX0+ID2FwMFu+nC1fzlq3ZkTM0ZF5e7OgIBYRwbKzH3l4QkKCo6MjEc2aNausrOzll18mIldX17t37xpeh48++oiI5s2bl5CQQEROTk5qtdrcF/bUKC8vd3BwIKIbN24wxpRKZdXUOTg4eHp6zp49OyIiIisr6+HDdTrd5MmT+a/96tWrdV79ekRYCGfPZj/8wMaNY8uXs/Xr2RdfsOhoZtSV7Pjx440aNSKi77//Pj8/v0ePHkTUr1+/kpISQw7XarWtW7cmovj4+KCgICL64IMPTHwxTyXeqvzSSy/xf6pUqsGDB8+aNWv79u0pKSmGlKDRaEaNGkVE7du3z8zMrMW61m+Cr4SzZ7PvvmOFhWzQIFMK2bZtm1wul8lkmzdvvn379nPPPUdEI0eONKRxJScnZ+jQoZ06dWKMde3alYh+++03UyrxtJo0aRIRLVq0yJxCSkpK+vfvT0TdunWrerNqiPz8/CNHjhQUFOh0OnPqIJzghplffmF2dmzpUtNLCA4OJiJLS8vDhw//+eeffEK94dc0tVp99epVInrmmWcqKipMr0d9VVxcXBvFqlQq/qvmC0yaIycnp1OnTkTk5eVVXl7+mJ9Uq9WJiYmbN2+eOnWqq6srX/TZ19d3+vTpZtZBLMEhHDqUETEzW6r/7//+jze+xcfHnzhxwtramoh++OEHAw//5ptviGjy5MlmVaLe0L9TAwICPD09raysvvrqK5O7cB7l4MGDRNS9e3dJSqva0qZvQWWMabXaP//8c9OmTR9++KGHh0e1xUcaNWrUrVs3S0tLIlpqzge5aCJDmJfHrKyYhQXLyTGrHJ1ON2HCBCJq3rx5UlJSeHi4m5tbamrqYw7RaDSXLl1av379tGnT+GSoXbt2mVUJcXQ63fXr17ds2TJz5sz+/fs3bty4Wq8dv84nJiZKeNL333+fiL7++mupCkxISODNPNOnTw8LC5s9e7aXl1e1NRPkcrmrq+ukSZNWrVrFl7FkjG3fvp0/kmzcuFGqytQxkSHcuJERsTfekKAolUr1xhtvENGLL76YlZXF/zzV6HuufH19q81ClMvlr7/+eo1H1U8ZGRn619K8efNq/QG8127JkiWnTp0qLi42uQvnUTQaDT/pn3/+KUmB3G+//WZpacmjWO218B7IRzV9r169mj+SHDx4UML61BmRIfT1ZURs3TppSissLOzVqxcReXh48AehnJycAwcOfP3118OHD+d9ylW1a9funXfeCQ4O3rZtG//uuHHj6v8jfllZmY+Pz8OpGzly5MKFCw8fPvxw80bVLpx79+6ZX4cjR44QEW/TMl/VbiH+pNe7d+/58+eHh4cb3mTKt/exsbGJiYmRpFZ1SVgINZoCpdLzl1++r6kPyUQZGRnt2rXjAXNxcan2TnVycho2bFhQUNC+ffuyH+yCVCqVfNHR2bNnS1ab2tGjR49nn33W1tbW09MzICBg8+bNhtxn5ufnd+/enYzpwnmMadOmEdGXX35pZjncggULunbtGhkZefv2bZlMZmNjY0Jjkk6n4621zZs3v3btmiQVqzMShTAyki1dyn74gX3wAZs8mR04wJ7U0nj37q9KJV275iVNBe5LSkrq0qWLm5sbvz9xd3fXv1Mff5U7duwYf+43vEWn7l2/fp2IHBwcysrKjD02LS2tTZs2crl85sx95gyP1Wq1rVq1IiITxgnWyNXVlYiOHDnCFzQYPXq0aeWoVCq+vY+Li8udO3ckqVvdkCiEc+cyxtinnzJb28rxZo6ObMIEFh7OHvF2uXHjTaWSsrP/JU0FqtBqtWfOnLl48aKxw19CQ0NlMplcLt+xY4fktZLEokWLiGjSpEmmHX7p0qXBg3cRMXNGyEZFRRFRhw4dTC+iCt4/1KxZM7Vaze+Zt23bZnJpJSUlfHuf3r17FxUVSVLDOiBRCL/4gjHG5s5lly+zb79lPXroR38WB76RnPxeXl6YVluq/3GttiQ+3kaplKtU6dJUQCJLliwhIisrq/rZcf/SSy8RkX46gglOnGDW1oyImdzHPmPGDCKaM2eOyXWo6uuvvyaiKVOm3LlzR6FQWFtbFxQUmFNgdnY2395n8ODBDaXjV6IQRkSw5cuZtzf7/HMWG8t0Onb9Olu8mLm7J+/srFSSUknx8Y2TknxzczdrNAX37v1XqaSrV1+W5uySCgwMJKImTZqcP39edF0ekJycLJPJ7O3tTbgXrSo8nCkUTCYzuntWo9GcP3+et4sqlUr+xbi4OHMqw59UDxw4wOf4vvnmm+aUxt24cYNv72NCS9vdu1uys9dkZn5nfjUMJ13DTHk5a9Kk8gLYti0LCGCnTjGttrw8+c6d5Veu9FMqZTyNSUk+BQW/Xbs2KCvrJ8nOLh2tVssXHW3durWBYyCrHV5YWFgL9WJLly4lovHjx5tf1Nq1jIgpFOyJE4mq9uvwEfPPPvusvb0975xYvHixTCb76ScT/478EdfR0bGiooLvG/nrr7+aVlQ1cXFxvKWtxit2eXl5SUlqUdHpe/d2ZmWF3L79eXLyP65d87p9e05W1sqcnJ/VaiOmAZhPuhBqtezkSTZzJmvTRn8vWjJrRFrax4WFv+t0moqKtKyskKtXB6amvs8/bAoLjyYljcjN3aTRSNBuLqGKigpvb2/e65j9mKkc91V9pzZr1mzcuHEff/yx5BMy+vTpQ0S7d++WpLS5cxkRc3Ji1R6dsrPZ/v0sKIj5+LDhw5dWa2F2cXFp3749ET3//PPp6elmPkXzR9yJEyfm5ORYWFhYWlpK0oPCHT16lLe0vfPOOwsWLJg8ebKPj0+3bt34lfzYMU9+Saj63/Xrr+t0FeXl11NT63T4VO10USQmsqAg1rFj6vau/OVduNAsJcU/Pz9Cp1PpP2xSU6fw7547p7h61TMrK0Slqi9D6QsKCnr27ElEffr0ebjFPC8v77ffflu4cOGoUaN4U2FV9vb2RDR58mQJex1v3brFm+/N72DgdDo2cyZ75x0WE8OWL2c7drB332Xt2z8wmdPd/fDDM5JKS0s9PT2JyM3N7d69e4sXL+ZP0UeOHDG2DvwRNyIigu/cOnz4cEleml5oaGijRo2qTfznzeZ79oy8cqXfjRtvpqV9nJm5KDd3c0HBb+XlN3JyfsnJWZeRIdlIIEPUbj9hSXHc7dtfXLr0ov7D5saNkfoPG7U6Kzt77fXrr587Z8G/++fxVszbm61ezerBxJb09PTnn3+eiHx9fUtLSxMTE9etW+fv768fOqzXpEkTfa9dcnJybGwsvxeaP3++VJVZsWIFEb377rtSFcjpJ3POmlUZPFtb9sorbNYstn07S0nR1nhUbm5uly5diOjVV18tLy837Sk6JSVF/4jLRztt2LBBmld1X3p6OhFZW1svWLBg3bp1kZGR8fHxmZmZj/9wLC6OuX59SE6ORINIDFBHnfWlpYnp6UGJia7Z2av4h01q6pSUlEn5+ZE6XblanZuTsz4pyefeQq/K94Jczjw9WUgIu327bmpYo8uXL/O5AtXW57O2tu7fv39AQMCWLVuuXbv28B91//79FhYWRPTjjz9KUhN+8ZG870Q/mTMujv3yC0tIYAZ2Id66datt27b8Zk+tVuufoh8/ZLeqZcuW8UfcvLw8KysrS0tLoyZkG4Iv5TxkyBCjjrp7d5tSSYmJHRmr+TNIcnU9YiY//3Bq6tTi4rM3b06732pqw1tNtdoilpfH/vMfNnJkZTs6bz0QelU8ffr0jBkzFAqFq6urv79/SEjIqVOnHj/jhvv111/585L5Q8MzMzPlcnnjxo0ln5e0fLnpkzkvXbrEh+BOnz5d/xTdoUMHQ56iGWMffPCBTCYLCwvbsGGDCVExxGeffUZECxYsMOoonU6dkPCcUkn5+Qckr1KN6jqEKSkTlErKzFxcXn49M3PR5cu99Heql0+2ZWPHsp07WXExKylhERHM3595e9dxDatRqVT/+te/THv3f/vtt0TUo8c/T50y6+Hwp59+IqK3337bnEIeplYzV1c2dSoz4COlZr///jtf3GDp0qWPf4qu0a1bt8rKyoYPH05EP//8s4mVeLRXXnmFiPbv32/sgXfuLOPtNJJXqUZ1GkKdTnX+/DNKJZWV/bWmSEVFalZWyNWrnrmLX6m8+llbM19ftnkzKyj4a0CcIAcPHly4cKHJh8+du9XOTvfMM8yciUSvvvoqEYWGhppeRE2OHGFEzMxh2FVnElV9ijawcbigoMDa2lqhUNS4Do05NBqNnZ2dTCbLMX6mnEZTcP58E6WSSkulGZr3eHUawvz8/UolXb7cs8bv6tJS2YoVbMAAJpf/lcaZMxm7PyJHhC5dupizBJtWy/z8GBFzdjZuBR09PpSkUaNGZg4ledi0aYyImT8Mu+pMouvXrzdv3jwgIKDq3NwapaWl7dq1i18GB5l2Q/xYFy5c4J1Mph2elvaxUkl101dRpyFMTZ2sVFJGxrdP+LnsbLZ5M/P1ZX37/jUgToR9+/bZ2tqaOc+wrIy9/DIjYl27MmO7wTIzM6dPn05EI0aMMKcOD9NqWatWjIhJMgy76kyi9PSahyIWFBScOnUqJCRkzJgxfB49N378eBO6N56Id3uYPLahvDxp507X9957VfJL9MPqLoQ6nfrCBSelksrKLht6jEpVOSBu5crarNojvfbaa7169TK/nPx81r07I2KvvPKoAe2VioqK+DuV94Xwt6mzs/PgwYPNr0ZVUVGMiEk0DPuvmUROTk76mUSlpaXR0dEhISHjx4/v0KFDtR0+mjZtOmTIkPnz5//xxx/SVOJBfD1Fk0fzMMbefPNNIvrqq68krFWN6i6ESUlRUVHNEhNd6+yMZsq6cKGZXD5y5EhJSktLqxxKtGTJA1+vqGCxsWzVKjZxInNz0/AOxqo9kB4eHrzxg6+xK5UZMxgRk2gYNmNVFjdwdnaeOHFir169+Ooveo0bNx4wYMDMmTNDQ0OvX78u2Ykfga+gZ07CT5w4QUQtWrQwc7DuE9VdCD/44AOFQrFuXYNZkKdw/Hi1XJ4o3Vv/0iX28ccsIIDFxLAFC9iMGaxvX9ao0QODVHr2fFXfF6JUKvmTVWRkJO91XCnRHYFWqx0x4kKzZlrzRl9Xx2cS8cmcvHPV2H4dqRQWFvIHaTNP6uHhURujCKqpoxBW3fWhbs5orsxMZm3N5HJ2f7qAVPgglYULK5uf5HLWpQubMIGtXMn++OORc6H//e9/87d1WFiY+XWIjo4mohde6Gh+UdXcu3fv8uXLwcHBJ06cqKXVFg1x7NgxIurbt6+Z5fz6669E5ObmFhgYOG/evIiIiNqYLlxHIay660PD8M47jIjZ2ZneifYI+kEqa9awY8dYfr6hB/Jlwhs3bnzq1Ckz6/DJJ58Q0WeffWZmOfXWd999R0QBAQFmlqNSqfjuQ/yJgGvXrt3YsWNXrFiRdvbsEx7xDVNHIdTv+lA3pzNXZiZzcGBErH17ycs2Z5AKn1DbrFmzy5cNbtx6iE6n4715Z8+eNbmQeo6vri9Jz+rQoUMdHR3nzJkzZ86cQYMG8dH5XHbv3szCgrm6Mn9/tm4dS0xkT+qYqVFdhLDqrg91cDoJLFhQeafo6yu6Kg/QaDRvv/02EbVp0yYtLc20Qs6ePctLqP9LyxmrsLAwJCRkxIgR/P3Gd6oxmVar5bcMMpls3rx5sbGxKpVKq9XyFWunTp2qevllplA88Ez/zDNs1SpjT1S7IeTr0n755Zf8Il6r55KMVsuaNmWWlkwuZxINv65Kvx2VaUpLSwcOHEj3ZxKZUAIfURkYGGhiDeql1NTUWbNmVV2z1MHBwZxPmYqKivfee4+IrKys+OI3dH/gfmBg4NatW28mJzPGWHExO3WKhYQwf3/m6sqImPFr5EgfwszMzIfXpXVxcWnVqpWEUzZr0a1brFu3yg+2Wrh0mxlC9uBMIsNbz4uLi0+cOBEcHMxHXUu+ML4oSqXS39+ftx4Tkaen54YNG6ZMmcJHz5mwNgJjrKioaMiQIURkZ2d36NChHTt2TJo0qdoUtlUDBjAHh7/28eOD427fZjt2GDvQUoIQFhYWRkVFLV261M/Pj09vqapVq1Y+Pj786daoN40wfLTqnDnMx4cx8+JSE/2ecObQzyQaNWrUo7agenhTCv4XadSo0YoVK544rKy+02pZRATz9p7ctSsRWVpajhkzJiYmRqvVRkREdO7cmW9JYmNj88033+i3lzVEZmYJH4neqlWrajMk+T5QCxcuHDFiRIKPzwM3ojIZ69yZBQZWju4yZqCliSFMTk5etWrVwx8PvH/Zy8tr9uzZYWFht27d4j9v7L5lIul/ieZfs2oiValVZxLpv5iUlBQaGhoYGDhgwIBqm1JYWlr26tVr2rRp69evbwAfhY+Rn8+Cg9nzz/N3/83Bg+fMmZOenp6fnx8cHMzbnIjon//8p7+/Px+m06ZNm82bNxtyd3rjBnvxRfbqq/teeOGFpKSkJ/x0tb3diZivrwkDLU0J4cWLF/lGSJyFhUW1/uWKiorY2NhVq1ZNnDjx008/5UclJibyN820adNMOGnd0f8SJblm1Sb9Nqne3t5Dhw7l84+r6tChw8ObjTdgycls5kxmb1958enUia1Zw0pK/ve//wUEBOjbLTt37rxmzRq+DsjJkyf55ghENGFC9MWLjys+JoY5OTEi1r8/y801uO+IU6mYUsni4kwYaGlKCPmovF69eq1cufLs2bN8UEJ6evqOHTv4nU/Vz+Cq7TH6fcu++65Ol5Qzjv6XWDtXQmlt3769SZMmfB00InJ0dPT29ubbpxg4ubYhWb26Mn6enmzHDqbRMKWS+fvHeHnpHwh37NhR7VZLq9X+/PPP3t6T+BTx6dNZbm4NZR8+XJnuN95gtbNc3iMZHUK1Wu3k5EREvKvq+PHjgwcPrraTjlwu79Kly4QJE1auXFlt8F54eLhCoZDJZLU9FOjpcffu3fXr1+/atUuqTZfqI/6gvnQpmz6dXbzIVCq2ZQtzd+eZ1LVoMe399x8/GKuoiAUFVQ4SdHRkS5Y8MDhp40ZmacmI2MSJrO735jI6hIcPHyairl278n8eOnRI3wCj38Iqt8aPmvvWrFnDH1EOHKij5QOgwdM/qOfns8WLmbNz5SWxRQsWFMQMHkp25UrlvrR8ZhlvIY6OZjJZ5dRKIV2nRodw6tSpVGVEf2Fh4d69ex81hexRvvjiC95ydebMGWMrAE8j/YN6ejqzsmJErGNHFhLCTFoA8siRyi69ESMqV3wMDGRr1khcZcPJGGNkMD72JTs7OyEhoVu3boYfWA1jbMqUKRs3bnRycjp9+jTfshzgkSIj6do1sramGTNoxQrq1o28venBCYpGUakoLIwyMoi//T/9VLKamsC4EEZFRQ0aNKhjx47Xrl0z88RqtXrUqFEHDx50cXGJjo6uOtUaoG4EB1PTppSURIsXi6yG/Mk/UkVYWBgR8UUmzWRpablr167+/fsnJycPGTKkoKDA/DIBjDVmDMXGCq6DEVdCnU7Xtm3bjIyM+Ph4fd+LmbKzsz09PW/cuDF06NADBw7IzLjBAGigjLgSRkdHZ2RktG/fXqoEElGLFi0OHz5sb2/fsmXL4uJiqYoFMMRXXxERzZ8vuBpGhFDCe1EiOnny5MWLF4moSZMmZWVl27Zt0+l0kpQM0LBYGPhzjLE9e/YQkZ+fnyQn/uSTT+Lj4w8fPpySkqLRaIYNG1atxx+gtpWXU0gIpaQIroahV8I//vgjLS2tbdu2fOkbM6WkpMTHx9vb27/yyiv8AitVtgEMZ21NgYHUvr3gahgaQv29qCRtJzt27CCiUaNGlZSUREVFWVpajhgxwvxiAYzCnwkXLhRcDUNDuHv3bpLueqW/+u3du1ej0Xh7ez88AwDgKWFQCC9fvpySkuLk5NSvXz8iSk5Onj59ulqtNu2Ut27dUiqVdnZ2Q4YMwb0ogEEh7NSp06BBg/Ly8vbv36/T6UaNGrV27dqpU6caNdpGb+fOnYwxX1/fioqKY8eOKRSKkSNHmlAOwN+DQSFUKBSenp5arfbdd9+NiYnZuHGjnZ3dpk2b5s2bZ8Ip9Ve/8PBwlUrl5eWlX4oG4Glk+FhvvnZos2bNrly5cvToUb5mSUhIiFEDxm/fvi2Xy21sbIqLi3ljzBqBw9cB6gEjQqjRaN566y26v+jlli1b+HbQRm2k/uOPPxKRn59fYWGhtbW1XC7PyMgwvtoAfx/GzScsLS319PQkIjc3t7y8vEWLFhGRlZXV0aNHDSyB72C8devW0NBQInrttdeMrzPA34rRk3pzc3M7d+7M81NeXj5z5kwiatKkyQXDNptcvXq1t7d3QUEBX0laqm2GABouUxZ60i96+e6776rVaj6atHXr1qmpqQaWUFJSYmtrK5fLjZ2SD/D3Y+K6o5cuXeIrfH344YdlZWX8JnPIkCEGHr5161YiGjhwoGlnB/g7MXQAdzVubm579uwZOnTo6tWr27dvHx4ePm3atODg4McckpGRce7cuXPnzkVHR588edLe3p7fkQI85Yxb3qKa7du3jxs3jjG2cePGiRMnVvtuVlZWXFxcbGws//+9e/eqfvc///mPn5+fjY2NyWcH+HswK4REtHr16o8++sjS0jIyMtLT0/PChQvn7rty5UrVwh0dHXv37u3p6enu7t6vXz900ANw5oaQiD799NPvv//exsamoqJCq9Xqv25vb+/u7u7h4dGnT58+ffrwvSgAoBoJQsgYmzlzZkVFxYYNGzp27Oju7u7u7j5w4MCePXsqFApJagnwNyZBCLmCgoJGjRrxrSYAwHCShRAATGPcuqMAIDmEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQDCEEEAwhBBAMIQQQLD/B/70U2vkighNAAAAAElFTkSuQmCC\n",
            "text/plain": [
              "<PIL.PngImagePlugin.PngImageFile image mode=RGB size=300x300 at 0x7F2645AF0FD0>"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vuh706D2SFXw"
      },
      "source": [
        "As you may have guessed, this SMILES string does not contain much of structural and chemical information present in the actual molecule representation. \n",
        "\n",
        "The smiles string is a string of symbols -- it is very different from the binding structure of the molecules.  One option that we have is to train machine learning models on the smiles strings directly.\n",
        "\n",
        "This has been found to not be the most effective way of proceeding. Another approach that people have thought of is to convert this to a \"chemical fingerprint\". This is a\n",
        " vector of 1's and 0's that captures the presence or absence of specific features, as determined by local arrangement of atoms in a molecule.  \n",
        "\n",
        "\n",
        "There are many algorithms that give you this finger print. One example is Extended-connectivity fingerprints (ECFP) scheme, which comes as `GetMorganFingerprintAsBitVec` in Rdkit. Here is an example of ECFP4. The number at the end refers to bond distances used for featurizing."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "KNYhVaVaSFXy",
        "scrolled": true,
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "8e7eaf66-c70f-44e7-a682-3dcd2c3a20a7"
      },
      "source": [
        "print('ECFP4')\n",
        "molecule = MolFromSmiles(smiles[0])\n",
        "ECFP2 = Chem.GetMorganFingerprintAsBitVect(molecule, 2).ToBitString()\n",
        "ECFP2arr = np.array(list(map(int, ECFP2)))\n",
        "print(ECFP2arr[:20], '...')\n",
        "print(ECFP2arr.shape)"
      ],
      "execution_count": 7,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "ECFP4\n",
            "[0 1 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0] ...\n",
            "(2048,)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "hLDF_gKMSFX5"
      },
      "source": [
        "Lots of Rdkit's featurization algorithms are ported in DeepChem. This is a deep learning for chemistry toolkit that was developed at Stanford.\n",
        "\n",
        "We can list the descriptors for molecules from RDKIT. Note that in addition to fingerprints, there are all sorts of purely chemical features, such as\n",
        "\n",
        "\n",
        "*   Number of Hydrogen acceptors\n",
        "*   Number of aromatic rings\n",
        "*   Number of radical electrons\n",
        "*   Number of rings in the molecule\n",
        "*   Partial charges\n",
        "\n",
        "\n",
        "And so forth.  When you are doing machine learning with complete ignorance of why the small molecule binds to something, you might as well be as general as you can be!!\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Uy0vsVrBSFX7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "96aa633a-5b2a-44a5-c3a4-44c2b7b4a3c4"
      },
      "source": [
        "from deepchem.feat import RDKitDescriptors   \n",
        "\n",
        "for descriptor in RDKitDescriptors.allowedDescriptors:\n",
        "    print(descriptor)\n"
      ],
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "NOCount\n",
            "Chi1\n",
            "NumAromaticCarbocycles\n",
            "EState_VSA9\n",
            "VSA_EState5\n",
            "EState_VSA7\n",
            "NumSaturatedRings\n",
            "HeavyAtomMolWt\n",
            "Chi2n\n",
            "NumHeteroatoms\n",
            "PEOE_VSA4\n",
            "Chi2v\n",
            "EState_VSA5\n",
            "Chi4v\n",
            "NumValenceElectrons\n",
            "Kappa2\n",
            "EState_VSA4\n",
            "HeavyAtomCount\n",
            "NumSaturatedCarbocycles\n",
            "Ipc\n",
            "MaxPartialCharge\n",
            "PEOE_VSA8\n",
            "SlogP_VSA7\n",
            "PEOE_VSA11\n",
            "SMR_VSA3\n",
            "SMR_VSA2\n",
            "VSA_EState3\n",
            "NHOHCount\n",
            "PEOE_VSA1\n",
            "Chi0\n",
            "VSA_EState1\n",
            "SlogP_VSA12\n",
            "SlogP_VSA2\n",
            "Chi1v\n",
            "SMR_VSA10\n",
            "FractionCSP3\n",
            "VSA_EState7\n",
            "MinAbsPartialCharge\n",
            "SMR_VSA8\n",
            "MolLogP\n",
            "PEOE_VSA9\n",
            "SMR_VSA5\n",
            "EState_VSA11\n",
            "EState_VSA2\n",
            "PEOE_VSA6\n",
            "ExactMolWt\n",
            "BertzCT\n",
            "RingCount\n",
            "PEOE_VSA14\n",
            "Chi1n\n",
            "PEOE_VSA5\n",
            "LabuteASA\n",
            "NumHAcceptors\n",
            "SlogP_VSA4\n",
            "MinPartialCharge\n",
            "Chi0n\n",
            "SMR_VSA7\n",
            "TPSA\n",
            "EState_VSA1\n",
            "NumAliphaticRings\n",
            "SlogP_VSA6\n",
            "NumRotatableBonds\n",
            "NumAromaticRings\n",
            "PEOE_VSA7\n",
            "EState_VSA8\n",
            "MinAbsEStateIndex\n",
            "Kappa1\n",
            "BalabanJ\n",
            "SlogP_VSA11\n",
            "HallKierAlpha\n",
            "SMR_VSA1\n",
            "SlogP_VSA1\n",
            "NumAliphaticHeterocycles\n",
            "Chi3v\n",
            "VSA_EState8\n",
            "SlogP_VSA3\n",
            "PEOE_VSA2\n",
            "VSA_EState9\n",
            "MolMR\n",
            "PEOE_VSA13\n",
            "EState_VSA3\n",
            "MaxAbsPartialCharge\n",
            "Chi4n\n",
            "SMR_VSA6\n",
            "Kappa3\n",
            "VSA_EState4\n",
            "NumAliphaticCarbocycles\n",
            "SlogP_VSA9\n",
            "SMR_VSA9\n",
            "MaxAbsEStateIndex\n",
            "VSA_EState2\n",
            "SMR_VSA4\n",
            "EState_VSA6\n",
            "SlogP_VSA5\n",
            "MolWt\n",
            "NumSaturatedHeterocycles\n",
            "Chi3n\n",
            "SlogP_VSA10\n",
            "PEOE_VSA12\n",
            "PEOE_VSA10\n",
            "NumHDonors\n",
            "NumRadicalElectrons\n",
            "SlogP_VSA8\n",
            "NumAromaticHeterocycles\n",
            "MinEStateIndex\n",
            "PEOE_VSA3\n",
            "EState_VSA10\n",
            "VSA_EState6\n",
            "Chi0v\n",
            "MaxEStateIndex\n",
            "VSA_EState10\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "NazROs-xSFYA"
      },
      "source": [
        "Deepchem has also implemented a large number of different machine learning models to build models on top of these features. We can list this as follows:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "SOnVGtOgSFYB",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "6e10e086-6dbf-418f-c013-9e6245b3a6cc"
      },
      "source": [
        "dir(dc.models)"
      ],
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "['ANIRegression',\n",
              " 'AtomicConvModel',\n",
              " 'BPSymmetryFunctionRegression',\n",
              " 'ChemCeption',\n",
              " 'DAGModel',\n",
              " 'DAGTensorGraph',\n",
              " 'DTNNModel',\n",
              " 'DTNNTensorGraph',\n",
              " 'GAN',\n",
              " 'GraphConvModel',\n",
              " 'GraphConvTensorGraph',\n",
              " 'IRV',\n",
              " 'KerasModel',\n",
              " 'MPNNModel',\n",
              " 'MPNNTensorGraph',\n",
              " 'Model',\n",
              " 'MultitaskClassifier',\n",
              " 'MultitaskFitTransformRegressor',\n",
              " 'MultitaskRegressor',\n",
              " 'OntologyModel',\n",
              " 'OntologyNode',\n",
              " 'ProgressiveMultitaskClassifier',\n",
              " 'ProgressiveMultitaskRegressor',\n",
              " 'RobustMultitaskClassifier',\n",
              " 'RobustMultitaskRegressor',\n",
              " 'ScScoreModel',\n",
              " 'SeqToSeq',\n",
              " 'SequenceDNN',\n",
              " 'Sequential',\n",
              " 'SingletaskToMultitask',\n",
              " 'SklearnModel',\n",
              " 'Smiles2Vec',\n",
              " 'TensorGraph',\n",
              " 'TensorflowMultitaskIRVClassifier',\n",
              " 'TextCNNModel',\n",
              " 'TextCNNTensorGraph',\n",
              " 'ValidationCallback',\n",
              " 'WGAN',\n",
              " 'WeaveModel',\n",
              " 'WeaveTensorGraph',\n",
              " 'XGBoostModel',\n",
              " '__builtins__',\n",
              " '__cached__',\n",
              " '__doc__',\n",
              " '__file__',\n",
              " '__loader__',\n",
              " '__name__',\n",
              " '__package__',\n",
              " '__path__',\n",
              " '__spec__',\n",
              " 'atomic_conv',\n",
              " 'callbacks',\n",
              " 'chemnet_models',\n",
              " 'create_gene_ontology',\n",
              " 'division',\n",
              " 'fcnet',\n",
              " 'gan',\n",
              " 'graph_models',\n",
              " 'keras_model',\n",
              " 'layers',\n",
              " 'losses',\n",
              " 'models',\n",
              " 'multitask',\n",
              " 'optimizers',\n",
              " 'progressive_multitask',\n",
              " 'robust_multitask',\n",
              " 'scscore',\n",
              " 'seqtoseq',\n",
              " 'sklearn_models',\n",
              " 'tensorgraph',\n",
              " 'text_cnn',\n",
              " 'unicode_literals',\n",
              " 'xgboost_models']"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "6Cl4Ai8CSFYE"
      },
      "source": [
        "## 2. Parsing KIBA dataset\n",
        "\n",
        "Now we are ready to examine the Kinase dataset.  The way this works is that we are given a particular protein, and for that protein, we want to predict the small molecules that bind to it. Different proteins of course have different sets of small molecules.  This dataset contains a large set of proteins, and for each of them, it contains the small molecules that bind.\n",
        "\n",
        "To associate proteins with small molecules, we also need a representation of the proteins. Proteins are of course sequences of amino acids, so the representation is just a string of amino acids.   To use the string of amino acids with numbers. We do this by simply mapping each amino acid (represented by a letter) to a number."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WTIiSHZRSFYF"
      },
      "source": [
        "import matplotlib.pyplot as plt\n",
        "import json\n",
        "import pickle\n",
        "from collections import OrderedDict\n",
        "import networkx as nx"
      ],
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Hu_V0gH7SFYK"
      },
      "source": [
        "# for converting protein sequence to categorical format\n",
        "seq_voc = \"ABCDEFGHIKLMNOPQRSTUVWXYZ\"\n",
        "seq_dict = {v:i for i,v in enumerate(seq_voc)}\n",
        "seq_dict_len = len(seq_dict)\n",
        "max_seq_len = 1000   # Note that all protein data will have the same length 1000 \n",
        "\n",
        "def seq_to_cat(prot):\n",
        "    x = np.zeros(max_seq_len)\n",
        "    for i, ch in enumerate(prot[:max_seq_len]): \n",
        "        x[i] = seq_dict[ch]\n",
        "    return x  \n",
        "\n",
        "# for Concordance index evaluation\n",
        "def ci(y,f):\n",
        "    ind = np.argsort(y)\n",
        "    y = y[ind]\n",
        "    f = f[ind]\n",
        "    i = len(y)-1\n",
        "    j = i-1\n",
        "    z = 0.0\n",
        "    S = 0.0\n",
        "    while i > 0:\n",
        "        while j >= 0:\n",
        "            if y[i] > y[j]:\n",
        "                z = z+1\n",
        "                u = f[i] - f[j]\n",
        "                if u > 0:\n",
        "                    S = S + 1\n",
        "                elif u == 0:\n",
        "                    S = S + 0.5\n",
        "            j = j - 1\n",
        "        i = i - 1\n",
        "        j = i - 1\n",
        "    ci = S/z\n",
        "    return ci if z != 0 else 0"
      ],
      "execution_count": 11,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "E_FG7L-Rn1oV"
      },
      "source": [
        "We read in the ligands and the proteins, as well as the binding data."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "5advqtV7SFYP",
        "scrolled": true
      },
      "source": [
        "fpath = G_PATH + 'data_Drug_target_binding_affinity/data/kiba/'\n",
        "\n",
        "# Read in drugs and proteins\n",
        "drugs_ = json.load(open(fpath + \"ligands_can.txt\"), object_pairs_hook=OrderedDict)\n",
        "drugs = np.array([Chem.MolToSmiles(Chem.MolFromSmiles(d),isomericSmiles=True) for d in drugs_.values()])\n",
        "proteins_ = json.load(open(fpath + \"proteins.txt\"), object_pairs_hook=OrderedDict)\n",
        "proteins = np.array(list(proteins_.values()))\n",
        "\n",
        "# Read in affinity data\n",
        "affinity = np.array(pickle.load(open(fpath + \"Y\",\"rb\"), encoding='latin1'))\n",
        "\n",
        "# Read in train/test fold  \n",
        "train_fold = json.load(open(fpath + \"folds/train_fold_setting1.txt\"))\n",
        "train_fold = [ee for e in train_fold for ee in e ]    \n",
        "'''\n",
        "Here all validation folds are aggregated into training set. \n",
        "If you want to train models with different architectures and/or \n",
        "optimize for model hyperparameters, we encourage you to use 5-fold \n",
        "cross validation as provided here.\n",
        "'''\n",
        "test_fold = json.load(open(fpath + \"folds/test_fold_setting1.txt\"))\n",
        "\n",
        "# Prepare train/test data with fold indices\n",
        "rows, cols = np.where(np.isnan(affinity)==False) \n",
        "drugs_tr = drugs[rows[train_fold]]    # (98545,)\n",
        "proteins_tr = np.array([seq_to_cat(p) for p in proteins[cols[train_fold]]])   # (98545, 1000)\n",
        "affinity_tr = affinity[rows[train_fold], cols[train_fold]]  # (98545,)\n",
        "\n",
        "drugs_ts = drugs[rows[test_fold]] # (19709,)\n",
        "proteins_ts = np.array([seq_to_cat(p) for p in proteins[cols[test_fold]]]) # (19709, 1000)\n",
        "affinity_ts = affinity[rows[test_fold], cols[test_fold]]    # (19709,)"
      ],
      "execution_count": 12,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "GudunykUcni_",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "f1b30ee9-9b8d-4938-a066-535b6e8c8b06"
      },
      "source": [
        "print('Example of drug:{}'.format(drugs_tr[0]))\n",
        "print('Example of protein:{} ...'.format(proteins_tr[0][:10]))\n",
        "print('Example of affinity score:{}'.format(affinity_tr[0]))"
      ],
      "execution_count": 13,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Example of drug:O=C1c2c(c3c4ccc(O)cc4n(C4OC(CO)C(O)C(O)C4O)c3c3[nH]c4cc(O)ccc4c23)C(=O)N1NC(CO)CO\n",
            "Example of protein:[11. 17.  0. 12. 12. 17. 14. 14. 17.  0.] ...\n",
            "Example of affinity score:9.798970004000001\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "w9_fbrsHPXh4",
        "outputId": "4785b986-15a2-432f-ff9d-b82789cb1d29"
      },
      "source": [
        "# Convert to ECFP fingerprint\n",
        "smileToMol = lambda x: MolFromSmiles(x)\n",
        "featurizer = dc.feat.CircularFingerprint(size=1024)\n",
        "\n",
        "drugs_mol_tr = list(map(smileToMol, drugs_tr))\n",
        "drugs_ecfp_tr = featurizer.featurize(drugs_mol_tr)\n",
        "drugs_mol_ts = list(map(smileToMol, drugs_ts))\n",
        "drugs_ecfp_ts = featurizer.featurize(drugs_mol_ts)\n",
        "\n",
        "print(drugs_ecfp_tr.shape)\n",
        "print(drugs_ecfp_ts.shape)"
      ],
      "execution_count": 33,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "(98545, 1024)\n",
            "(19709, 1024)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "QQbj9m7WldB6"
      },
      "source": [
        "tr_size, drug_size = drugs_ecfp_tr.shape[0], drugs_ecfp_tr.shape[1]\n",
        "ts_size = drugs_ecfp_ts.shape[0]\n",
        "\n",
        "protein_size = max_seq_len"
      ],
      "execution_count": 132,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "GRXYpyYVSFYS"
      },
      "source": [
        "## 3. Train a model on KIBA data"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "qODLvdvESFYT",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "ff7bd961-4e7a-4e42-ec09-a38ace1aeafa"
      },
      "source": [
        "# Have fun!\n",
        "import tensorflow as tf\n",
        "from keras import Sequential, Model\n",
        "from keras.layers import Dense, concatenate, Dropout\n",
        "\n",
        "def build_baseline_model(drug_size, protein_size):\n",
        "\n",
        "  drug_model = Sequential()\n",
        "  drug_model.add(Dense(1, input_shape=(drug_size,), activation='linear'))\n",
        "\n",
        "  protein_model = Sequential()\n",
        "  protein_model.add(Dense(1, input_shape=(protein_size,), activation='linear'))\n",
        "\n",
        "  # concat_layer = tf.keras.layers.Concatenate([drug_model.outputs[0], protein_model.outputs[0]])\n",
        "  model_concat = concatenate([drug_model.output, protein_model.output])\n",
        "  # fully connected\n",
        "  model_concat = Dense(1024, activation='relu')(model_concat)\n",
        "  model_concat = Dropout(0.1)(model_concat)\n",
        "  model_concat = Dense(1024, activation='relu')(model_concat)\n",
        "  model_concat = Dropout(0.1)(model_concat)\n",
        "  model_concat = Dense(512, activation='relu')(model_concat)\n",
        "  model_concat = Dense(1, kernel_initializer='normal')(model_concat)\n",
        "  \n",
        "  model = Model(inputs=[drug_model.input, protein_model.input], outputs=model_concat)\n",
        "\n",
        "  model.compile(optimizer='adam', loss='mean_squared_error', metrics=['mean_squared_error'])\n",
        "  \n",
        "  return model\n",
        "\n",
        "model = build_baseline_model(drug_size=drug_size, protein_size=protein_size)\n",
        "print(model.summary())"
      ],
      "execution_count": 133,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Model: \"model_17\"\n",
            "__________________________________________________________________________________________________\n",
            "Layer (type)                    Output Shape         Param #     Connected to                     \n",
            "==================================================================================================\n",
            "dense_113_input (InputLayer)    (None, 1024)         0                                            \n",
            "__________________________________________________________________________________________________\n",
            "dense_114_input (InputLayer)    (None, 1000)         0                                            \n",
            "__________________________________________________________________________________________________\n",
            "dense_113 (Dense)               (None, 1)            1025        dense_113_input[0][0]            \n",
            "__________________________________________________________________________________________________\n",
            "dense_114 (Dense)               (None, 1)            1001        dense_114_input[0][0]            \n",
            "__________________________________________________________________________________________________\n",
            "concatenate_32 (Concatenate)    (None, 2)            0           dense_113[0][0]                  \n",
            "                                                                 dense_114[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "dense_115 (Dense)               (None, 1024)         3072        concatenate_32[0][0]             \n",
            "__________________________________________________________________________________________________\n",
            "dropout_13 (Dropout)            (None, 1024)         0           dense_115[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "dense_116 (Dense)               (None, 1024)         1049600     dropout_13[0][0]                 \n",
            "__________________________________________________________________________________________________\n",
            "dropout_14 (Dropout)            (None, 1024)         0           dense_116[0][0]                  \n",
            "__________________________________________________________________________________________________\n",
            "dense_117 (Dense)               (None, 512)          524800      dropout_14[0][0]                 \n",
            "__________________________________________________________________________________________________\n",
            "dense_118 (Dense)               (None, 1)            513         dense_117[0][0]                  \n",
            "==================================================================================================\n",
            "Total params: 1,580,011\n",
            "Trainable params: 1,580,011\n",
            "Non-trainable params: 0\n",
            "__________________________________________________________________________________________________\n",
            "None\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "ViStDljpY7iN",
        "outputId": "d844d010-adb0-4050-f049-e5e1edaaba89"
      },
      "source": [
        "from keras.callbacks import EarlyStopping\n",
        "\n",
        "epochs = 100\n",
        "batch_size = 50\n",
        "\n",
        "callback = EarlyStopping(monitor='val_loss', patience=5, restore_best_weights=True)\n",
        "\n",
        "train_history = model.fit([drugs_ecfp_tr, proteins_tr], affinity_tr, \n",
        "                    validation_split=0.2, batch_size=batch_size, epochs=epochs, verbose=True)\n"
      ],
      "execution_count": 135,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Train on 78836 samples, validate on 19709 samples\n",
            "Epoch 1/100\n",
            "78836/78836 [==============================] - 68s 865us/step - loss: 0.5321 - mean_squared_error: 0.5321 - val_loss: 2.0868 - val_mean_squared_error: 2.0868\n",
            "Epoch 2/100\n",
            "78836/78836 [==============================] - 69s 873us/step - loss: 0.5283 - mean_squared_error: 0.5283 - val_loss: 2.1565 - val_mean_squared_error: 2.1565\n",
            "Epoch 3/100\n",
            "78836/78836 [==============================] - 67s 853us/step - loss: 0.5310 - mean_squared_error: 0.5310 - val_loss: 1.2997 - val_mean_squared_error: 1.2997\n",
            "Epoch 4/100\n",
            "78836/78836 [==============================] - 66s 838us/step - loss: 0.5190 - mean_squared_error: 0.5190 - val_loss: 1.6997 - val_mean_squared_error: 1.6997\n",
            "Epoch 5/100\n",
            "78836/78836 [==============================] - 67s 844us/step - loss: 0.5052 - mean_squared_error: 0.5052 - val_loss: 1.7902 - val_mean_squared_error: 1.7902\n",
            "Epoch 6/100\n",
            "78836/78836 [==============================] - 67s 855us/step - loss: 0.5030 - mean_squared_error: 0.5030 - val_loss: 0.9893 - val_mean_squared_error: 0.9893\n",
            "Epoch 7/100\n",
            "78836/78836 [==============================] - 68s 862us/step - loss: 0.4863 - mean_squared_error: 0.4863 - val_loss: 1.1108 - val_mean_squared_error: 1.1108\n",
            "Epoch 8/100\n",
            "78836/78836 [==============================] - 68s 864us/step - loss: 0.4741 - mean_squared_error: 0.4741 - val_loss: 1.1121 - val_mean_squared_error: 1.1121\n",
            "Epoch 9/100\n",
            "78836/78836 [==============================] - 69s 875us/step - loss: 0.4679 - mean_squared_error: 0.4679 - val_loss: 1.2602 - val_mean_squared_error: 1.2602\n",
            "Epoch 10/100\n",
            "78836/78836 [==============================] - 75s 950us/step - loss: 0.4689 - mean_squared_error: 0.4689 - val_loss: 0.7032 - val_mean_squared_error: 0.7032\n",
            "Epoch 11/100\n",
            "78836/78836 [==============================] - 69s 874us/step - loss: 0.4634 - mean_squared_error: 0.4634 - val_loss: 0.8672 - val_mean_squared_error: 0.8672\n",
            "Epoch 12/100\n",
            "78836/78836 [==============================] - 68s 869us/step - loss: 0.4532 - mean_squared_error: 0.4532 - val_loss: 1.0899 - val_mean_squared_error: 1.0899\n",
            "Epoch 13/100\n",
            "78836/78836 [==============================] - 68s 865us/step - loss: 0.4539 - mean_squared_error: 0.4539 - val_loss: 0.7825 - val_mean_squared_error: 0.7825\n",
            "Epoch 14/100\n",
            "78836/78836 [==============================] - 70s 892us/step - loss: 0.4447 - mean_squared_error: 0.4447 - val_loss: 0.7623 - val_mean_squared_error: 0.7623\n",
            "Epoch 15/100\n",
            "78836/78836 [==============================] - 70s 888us/step - loss: 0.4423 - mean_squared_error: 0.4423 - val_loss: 0.6661 - val_mean_squared_error: 0.6661\n",
            "Epoch 16/100\n",
            "78836/78836 [==============================] - 71s 899us/step - loss: 0.4376 - mean_squared_error: 0.4376 - val_loss: 0.7398 - val_mean_squared_error: 0.7398\n",
            "Epoch 17/100\n",
            "78836/78836 [==============================] - 69s 874us/step - loss: 0.4366 - mean_squared_error: 0.4366 - val_loss: 0.5680 - val_mean_squared_error: 0.5680\n",
            "Epoch 18/100\n",
            "78836/78836 [==============================] - 69s 872us/step - loss: 0.4361 - mean_squared_error: 0.4361 - val_loss: 0.5950 - val_mean_squared_error: 0.5950\n",
            "Epoch 19/100\n",
            "78836/78836 [==============================] - 69s 881us/step - loss: 0.4388 - mean_squared_error: 0.4388 - val_loss: 0.5177 - val_mean_squared_error: 0.5177\n",
            "Epoch 20/100\n",
            "78836/78836 [==============================] - 70s 882us/step - loss: 0.4268 - mean_squared_error: 0.4268 - val_loss: 0.4211 - val_mean_squared_error: 0.4211\n",
            "Epoch 21/100\n",
            "78836/78836 [==============================] - 69s 876us/step - loss: 0.4268 - mean_squared_error: 0.4268 - val_loss: 0.4037 - val_mean_squared_error: 0.4037\n",
            "Epoch 22/100\n",
            "78836/78836 [==============================] - 70s 885us/step - loss: 0.4245 - mean_squared_error: 0.4245 - val_loss: 0.4666 - val_mean_squared_error: 0.4666\n",
            "Epoch 23/100\n",
            "78836/78836 [==============================] - 70s 883us/step - loss: 0.4260 - mean_squared_error: 0.4260 - val_loss: 0.4347 - val_mean_squared_error: 0.4347\n",
            "Epoch 24/100\n",
            "78836/78836 [==============================] - 73s 928us/step - loss: 0.4209 - mean_squared_error: 0.4209 - val_loss: 0.4373 - val_mean_squared_error: 0.4373\n",
            "Epoch 25/100\n",
            "78836/78836 [==============================] - 73s 920us/step - loss: 0.4202 - mean_squared_error: 0.4202 - val_loss: 0.4181 - val_mean_squared_error: 0.4181\n",
            "Epoch 26/100\n",
            "78836/78836 [==============================] - 74s 944us/step - loss: 0.4185 - mean_squared_error: 0.4185 - val_loss: 0.3985 - val_mean_squared_error: 0.3985\n",
            "Epoch 27/100\n",
            "78836/78836 [==============================] - 73s 930us/step - loss: 0.4185 - mean_squared_error: 0.4185 - val_loss: 0.4133 - val_mean_squared_error: 0.4133\n",
            "Epoch 28/100\n",
            "78836/78836 [==============================] - 73s 930us/step - loss: 0.4172 - mean_squared_error: 0.4172 - val_loss: 0.4011 - val_mean_squared_error: 0.4011\n",
            "Epoch 29/100\n",
            "78836/78836 [==============================] - 71s 899us/step - loss: 0.4152 - mean_squared_error: 0.4152 - val_loss: 0.4084 - val_mean_squared_error: 0.4084\n",
            "Epoch 30/100\n",
            "78836/78836 [==============================] - 74s 935us/step - loss: 0.4126 - mean_squared_error: 0.4126 - val_loss: 0.3969 - val_mean_squared_error: 0.3969\n",
            "Epoch 31/100\n",
            "78836/78836 [==============================] - 73s 921us/step - loss: 0.4150 - mean_squared_error: 0.4150 - val_loss: 0.4252 - val_mean_squared_error: 0.4252\n",
            "Epoch 32/100\n",
            "78836/78836 [==============================] - 73s 930us/step - loss: 0.4128 - mean_squared_error: 0.4128 - val_loss: 0.3986 - val_mean_squared_error: 0.3986\n",
            "Epoch 33/100\n",
            "78836/78836 [==============================] - 75s 951us/step - loss: 0.4123 - mean_squared_error: 0.4123 - val_loss: 0.3902 - val_mean_squared_error: 0.3902\n",
            "Epoch 34/100\n",
            "78836/78836 [==============================] - 75s 947us/step - loss: 0.4118 - mean_squared_error: 0.4118 - val_loss: 0.4294 - val_mean_squared_error: 0.4294\n",
            "Epoch 35/100\n",
            "78836/78836 [==============================] - 74s 942us/step - loss: 0.4136 - mean_squared_error: 0.4136 - val_loss: 0.3912 - val_mean_squared_error: 0.3912\n",
            "Epoch 36/100\n",
            "78836/78836 [==============================] - 74s 939us/step - loss: 0.4073 - mean_squared_error: 0.4073 - val_loss: 0.3938 - val_mean_squared_error: 0.3938\n",
            "Epoch 37/100\n",
            "78836/78836 [==============================] - 73s 922us/step - loss: 0.4093 - mean_squared_error: 0.4093 - val_loss: 0.4070 - val_mean_squared_error: 0.4070\n",
            "Epoch 38/100\n",
            "78836/78836 [==============================] - 70s 888us/step - loss: 0.4071 - mean_squared_error: 0.4071 - val_loss: 0.3885 - val_mean_squared_error: 0.3885\n",
            "Epoch 39/100\n",
            "78836/78836 [==============================] - 70s 889us/step - loss: 0.4057 - mean_squared_error: 0.4057 - val_loss: 0.4346 - val_mean_squared_error: 0.4346\n",
            "Epoch 40/100\n",
            "78836/78836 [==============================] - 70s 888us/step - loss: 0.4079 - mean_squared_error: 0.4079 - val_loss: 0.3941 - val_mean_squared_error: 0.3941\n",
            "Epoch 41/100\n",
            "78836/78836 [==============================] - 69s 877us/step - loss: 0.4059 - mean_squared_error: 0.4059 - val_loss: 0.4386 - val_mean_squared_error: 0.4386\n",
            "Epoch 42/100\n",
            "78836/78836 [==============================] - 70s 884us/step - loss: 0.4014 - mean_squared_error: 0.4014 - val_loss: 0.4010 - val_mean_squared_error: 0.4010\n",
            "Epoch 43/100\n",
            "78836/78836 [==============================] - 71s 899us/step - loss: 0.4038 - mean_squared_error: 0.4038 - val_loss: 0.3840 - val_mean_squared_error: 0.3840\n",
            "Epoch 44/100\n",
            "78836/78836 [==============================] - 71s 902us/step - loss: 0.4022 - mean_squared_error: 0.4022 - val_loss: 0.4147 - val_mean_squared_error: 0.4147\n",
            "Epoch 45/100\n",
            "78836/78836 [==============================] - 72s 908us/step - loss: 0.4026 - mean_squared_error: 0.4026 - val_loss: 0.4169 - val_mean_squared_error: 0.4169\n",
            "Epoch 46/100\n",
            "78836/78836 [==============================] - 72s 912us/step - loss: 0.4045 - mean_squared_error: 0.4045 - val_loss: 0.3893 - val_mean_squared_error: 0.3893\n",
            "Epoch 47/100\n",
            "78836/78836 [==============================] - 71s 904us/step - loss: 0.4027 - mean_squared_error: 0.4027 - val_loss: 0.3894 - val_mean_squared_error: 0.3894\n",
            "Epoch 48/100\n",
            "78836/78836 [==============================] - 70s 886us/step - loss: 0.3981 - mean_squared_error: 0.3981 - val_loss: 0.3937 - val_mean_squared_error: 0.3937\n",
            "Epoch 49/100\n",
            "78836/78836 [==============================] - 70s 884us/step - loss: 0.4020 - mean_squared_error: 0.4020 - val_loss: 0.3835 - val_mean_squared_error: 0.3835\n",
            "Epoch 50/100\n",
            "78836/78836 [==============================] - 69s 878us/step - loss: 0.4016 - mean_squared_error: 0.4016 - val_loss: 0.3818 - val_mean_squared_error: 0.3818\n",
            "Epoch 51/100\n",
            "78836/78836 [==============================] - 70s 884us/step - loss: 0.3981 - mean_squared_error: 0.3981 - val_loss: 0.4069 - val_mean_squared_error: 0.4069\n",
            "Epoch 52/100\n",
            "78836/78836 [==============================] - 76s 958us/step - loss: 0.4037 - mean_squared_error: 0.4037 - val_loss: 0.3813 - val_mean_squared_error: 0.3813\n",
            "Epoch 53/100\n",
            "78836/78836 [==============================] - 77s 972us/step - loss: 0.3974 - mean_squared_error: 0.3974 - val_loss: 0.3818 - val_mean_squared_error: 0.3818\n",
            "Epoch 54/100\n",
            "78836/78836 [==============================] - 76s 961us/step - loss: 0.3990 - mean_squared_error: 0.3990 - val_loss: 0.3858 - val_mean_squared_error: 0.3858\n",
            "Epoch 55/100\n",
            "16950/78836 [=====>........................] - ETA: 58s - loss: 0.4072 - mean_squared_error: 0.4072"
          ],
          "name": "stdout"
        },
        {
          "output_type": "error",
          "ename": "KeyboardInterrupt",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-135-7683fedba07a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m history = model.fit([drugs_ecfp_tr, proteins_tr], affinity_tr, \n\u001b[0;32m----> 9\u001b[0;31m                     validation_split=0.2, batch_size=batch_size, epochs=epochs, verbose=True)\n\u001b[0m",
            "\u001b[0;32m/tensorflow-1.15.2/python3.7/keras/engine/training.py\u001b[0m in \u001b[0;36mfit\u001b[0;34m(self, x, y, batch_size, epochs, verbose, callbacks, validation_split, validation_data, shuffle, class_weight, sample_weight, initial_epoch, steps_per_epoch, validation_steps, validation_freq, max_queue_size, workers, use_multiprocessing, **kwargs)\u001b[0m\n\u001b[1;32m   1237\u001b[0m                                         \u001b[0msteps_per_epoch\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msteps_per_epoch\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1238\u001b[0m                                         \u001b[0mvalidation_steps\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mvalidation_steps\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1239\u001b[0;31m                                         validation_freq=validation_freq)\n\u001b[0m\u001b[1;32m   1240\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1241\u001b[0m     def evaluate(self,\n",
            "\u001b[0;32m/tensorflow-1.15.2/python3.7/keras/engine/training_arrays.py\u001b[0m in \u001b[0;36mfit_loop\u001b[0;34m(model, fit_function, fit_inputs, out_labels, batch_size, epochs, verbose, callbacks, val_function, val_inputs, shuffle, initial_epoch, steps_per_epoch, validation_steps, validation_freq)\u001b[0m\n\u001b[1;32m    194\u001b[0m                     \u001b[0mins_batch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mins_batch\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtoarray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    195\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 196\u001b[0;31m                 \u001b[0mouts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfit_function\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mins_batch\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    197\u001b[0m                 \u001b[0mouts\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mto_list\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mouts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    198\u001b[0m                 \u001b[0;32mfor\u001b[0m \u001b[0ml\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mo\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mzip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mout_labels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mouts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/tensorflow-1.15.2/python3.7/tensorflow_core/python/keras/backend.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, inputs)\u001b[0m\n\u001b[1;32m   3474\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3475\u001b[0m     fetched = self._callable_fn(*array_vals,\n\u001b[0;32m-> 3476\u001b[0;31m                                 run_metadata=self.run_metadata)\n\u001b[0m\u001b[1;32m   3477\u001b[0m     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_call_fetch_callbacks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfetched\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_fetches\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3478\u001b[0m     output_structure = nest.pack_sequence_as(\n",
            "\u001b[0;32m/tensorflow-1.15.2/python3.7/tensorflow_core/python/client/session.py\u001b[0m in \u001b[0;36m__call__\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   1470\u001b[0m         ret = tf_session.TF_SessionRunCallable(self._session._session,\n\u001b[1;32m   1471\u001b[0m                                                \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_handle\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1472\u001b[0;31m                                                run_metadata_ptr)\n\u001b[0m\u001b[1;32m   1473\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mrun_metadata\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1474\u001b[0m           \u001b[0mproto_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtf_session\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mTF_GetBuffer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrun_metadata_ptr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Ak5RPuFCASFn"
      },
      "source": [
        "# save the trained model\n",
        "import os\n",
        "save_folder = f'{G_PATH}/saved_models/'\n",
        "os.makedirs(save_folder, exist_ok=True)\n",
        "save_path = save_folder + 'baseline_model'\n",
        "model.save(save_path)\n",
        "\n",
        "# load model\n",
        "# model = keras.models.load_model(save_path)"
      ],
      "execution_count": 137,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "p4HOQGJKIiBt",
        "outputId": "a861821b-20be-4360-c2f5-96a72205f9ee"
      },
      "source": [
        "# summarize history for loss\n",
        "# print(history.history.keys())\n",
        "# plt.plot(train_history.history['loss'])\n",
        "# plt.plot(train_history.history['val_loss'])\n",
        "# plt.title('model loss')\n",
        "# plt.ylabel('loss')\n",
        "# plt.xlabel('epoch')\n",
        "# plt.legend(['train', 'test'], loc='upper left')\n",
        "# plt.show()\n",
        "\n",
        "print(\"Evaluate on test data\")\n",
        "results = model.evaluate([drugs_ecfp_ts, proteins_ts], affinity_ts, batch_size=128)\n",
        "print(\"test MSE loss is:\", results)"
      ],
      "execution_count": 147,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Evaluate on test data\n",
            "19709/19709 [==============================] - 2s 105us/step\n",
            "test MSE loss is: [0.398003730737514, 0.3980037271976471]\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "y_CH3ziCNDux"
      },
      "source": [
        "predicted_affinity = model.predict([drugs_ecfp_ts, proteins_ts])"
      ],
      "execution_count": 143,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tT2skwaRPpm8",
        "outputId": "31a83482-b780-487d-b89f-06d7581d00d5"
      },
      "source": [
        "ci_score = ci(affinity_ts, predicted_affinity)\n",
        "print('test CI score is:', ci_score)"
      ],
      "execution_count": 161,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "test CI score is: 0.7536044611424282\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "K67QchKnSFYZ"
      },
      "source": [
        "[The original paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4364066/) obtained CI score of 0.782 and MSE of 0.411, and [random forest model](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5395521/) got 0.836 and 0.222. Can you beat them?"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "my0BqyOkSFYa"
      },
      "source": [
        "## 4. Use  your model on COVID-19 protease"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "rkmmMhjOSFYc"
      },
      "source": [
        "Now, use your trained model to identify drugs that could be used as COVID-19 protease inhibitors from these 2111 drugs in the dataset. The sequence of the protease is provided in `6Y84_A.fasta.txt`. You might want to first predict a binding affinity of Ritonavior, a well known HIV drug that binds to HIV protease, to get the sense of a good binding score for this task. SMILES of Ritonavior is provided below.  "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "DFt_emuuSFYd"
      },
      "source": [
        "ritonavior = 'CC(C)C1=NC(=CS1)CN(C)C(=O)NC(C(C)C)C(=O)NC(CC2=CC=CC=C2)CC(C(CC3=CC=CC=C3)NC(=O)OCC4=CN=CS4)O'\n",
        "rit_mol = MolFromSmiles(ritonavior)\n",
        "Draw.MolToImage(rit_mol)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "2meDhKVKSFYi"
      },
      "source": [
        "Please keep in mind that proteins in KIBA data are kinase family, a different kind from protease. So in this exercise, we are relying on transfer learning. That is, the learned embedding for predicting kinase could be useful for predicting protease binding as well."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dRd6CYBhe_-U"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}